{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stress molecules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This program was created as a part of my master's thesis that deals with a relationship between interspecies gene variants of cellular stress response and longevity.\n",
    "\n",
    "### It consists of:\n",
    "<ol>\n",
    "    <li> <strong>Get a list of stress molecule symbols</strong>\n",
    "        <ul>\n",
    "            <li> downloading list of signaling molecules of pathways from Reactome database selected by user</li>\n",
    "        </ul>\n",
    "    </li>\n",
    "\n",
    "<li><strong>Get orthologs for every stress molecule; from NCBI</strong>\n",
    "    <ul>\n",
    "        <li> sequence acquisition of orthologs of these molecules, i.e. genes, from NCBI databases (= fasta files)\n",
    "        <li> parsing the fasta files and creating of various tables with information about these genes an its sequences\n",
    "        <li> checking of for how many organisms the data has been downloaded and selecting the data that upcoming analysis will be performed with\n",
    "    </ul>\n",
    "</li>\n",
    "    \n",
    "    \n",
    "<li><strong>Animal Traits & AnAge databases; calculating longevity quotient</strong>\n",
    "    <ul>\n",
    "        <li> merging available datasets of characteristics of organisms\n",
    "        <li> calculationg longevity quotient\n",
    "        <li> selecting short- and long-living organisms\n",
    "    </ul>\n",
    "</li>\n",
    "    \n",
    "<li><strong>SHORT-living; Select reference sequences</strong>\n",
    "    <ul>\n",
    "        <li> selecting reference sequences for short-living organisms\n",
    "    </ul>\n",
    "</li>\n",
    "    \n",
    "<li><strong>LONG-living; Select reference sequences</strong>\n",
    "    <ul>\n",
    "        <li> selecting reference sequences for long-living organisms\n",
    "    </ul>\n",
    "</li>\n",
    "    \n",
    "<li><strong>Create files for CAASTools Discovery analysis</strong>\n",
    "    <ul>\n",
    "        <li> creating fasta files for multiple sequence alignment (MSA) and a configuration file with short- and long-living organisms\n",
    "        <li> performing MSA\n",
    "    </ul>\n",
    "</li> \n",
    "    \n",
    "<li><strong>Create csv files</strong>\n",
    "    <ul>\n",
    "        <li> creating csv files sith results from previous analyses\n",
    "    </ul>\n",
    "</li>\n",
    "    \n",
    "\n",
    "<li> Additional: Divide files from NCBI Datasets into subparts and select ref seq if NCBI connection error\n",
    "    <ul>\n",
    "        <li> sometimes an error occur, and it was not unfortunatelly find out why it happens, for that reaseon, there is slower running script, that may help)\n",
    "    </ul>\n",
    "</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Type in the path of the parental folder (for \"fpilarova_masters-main\" it would be C:\\\\Users\\\\frant\\\\Desktop\\\\fpilarova_masters-main):\n",
      " C:\\\\Users\\\\frant\\\\Desktop\\\\scripts_dp\n"
     ]
    }
   ],
   "source": [
    "win_path = str(input('Type in the path of the parental folder (for \"fpilarova_masters-main\" it would be C:\\\\\\\\Users\\\\\\\\frant\\\\\\\\Desktop\\\\\\\\fpilarova_masters-main):\\n'))\n",
    "\n",
    "main_path = win_path.replace('\\\\\\\\', '\\\\')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import httplib2 as http\n",
    "import json\n",
    "import os\n",
    "import csv\n",
    "import re\n",
    "import math\n",
    "import pandas as pd\n",
    "import fastaparser\n",
    "from collections import Counter\n",
    "from urllib.request import urlopen\n",
    "import requests\n",
    "from itertools import islice\n",
    "\n",
    "\n",
    "\n",
    "try:\n",
    "    from urlparse import urlparse\n",
    "except ImportError:\n",
    "    from urllib.parse import urlparse\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(action=\"ignore\", message=\"unclosed\", category=ResourceWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1 Get a list of stress molecule symbols\n",
    "**Description of the script:**\n",
    "1. List of gene symbols is downloaded; for every pathway requested (input: Reactome ID)\n",
    "2. The symbols are checked to be HGNC-approved\n",
    "    - if not -> printed (EDIT - ask HGNC for HGNC-approved name of that gene)\n",
    "3. Checked symbols are added to the final list (stress_molecules_list = containing stress molecules of all requested Reactome IDs)\n",
    "5. Table gene_in_multiple_paths_check is created; with information about in which Reactome pathways the genes are involved"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "stress_molecules_list = []\n",
    "#stress_molecules_list = ['AAAS', 'HSF1'] #TEST\n",
    "# final list of all stress molecules; to be saved as stress_molecules.csv\n",
    "      \n",
    "reactome_ids_symbols_dictionary = {}\n",
    "# dictionary of Reactome ids and their HGNC-approved gene symbols\n",
    "# {Reactome_ID_1: [gene_1, gene_2], ...}\n",
    "\n",
    "gene_in_multiple_paths_check = {}\n",
    "# dictionary of genes and Reactome IDs in which the genes are included\n",
    "# {gene_1: [Reactome_ID_1, Reactome_ID_2], ...} \n",
    "\n",
    "HGNC_not = []\n",
    "# list of gene symbols that are not HGNC-approved"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Ask user for Reactome path IDs\n",
    "**Input:**\n",
    "- reactome_ids = list of Reactome IDs\n",
    "- reactome_names (optional)\n",
    "\n",
    "\n",
    "**Testing input:**\n",
    "\n",
    "- R-HSA-3371556 R-HSA-9612973 R-HSA-9711097 R-HSA-381119 R-HSA-73894\n",
    "- hsr autophagy starvation upr dna-repair\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      " Enter Reactome path IDs (with spaces as separators): R-HSA-3371556 R-HSA-9612973 R-HSA-9711097 R-HSA-381119 R-HSA-73894\n",
      "\n",
      " Do you want to name them? \n",
      " Yes - enter the names (with spaces as separators) \n",
      " No - press enter \n",
      " hsr autophagy starvation upr dna-repair\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " List of Reactome IDs saved to be examined later:\n",
      "R-HSA-3371556 (hsr)\n",
      "R-HSA-9612973 (autophagy)\n",
      "R-HSA-9711097 (starvation)\n",
      "R-HSA-381119 (upr)\n",
      "R-HSA-73894 (dna-repair)\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ask for input\n",
    "\n",
    "reactome_ids_input = str(input(' Enter Reactome path IDs (with spaces as separators):')) # ask for IDs\n",
    "reactome_ids = reactome_ids_input.split(' ')\n",
    "pathway_names_input = str(input('\\n Do you want to name them? \\n Yes - enter the names (with spaces as separators) \\n No - press enter \\n')) #ask for names if needed\n",
    "pathway_names = pathway_names_input.split(' ')\n",
    "\n",
    "while (len(reactome_ids) != len(pathway_names)) and (pathway_names[0] != ''): # check if there is the same number of names and ids\n",
    "    print(' Error: there is different number of path IDs and path names.')\n",
    "    reactome_ids_input = str(input(' Enter Reactome path IDs (with spaces as separators): '))\n",
    "    reactome_ids = reactome_ids_input.split(' ')\n",
    "    pathway_names_input = str(input('Do you want to name them? \\n Yes - enter the names (with spaces as separators) \\n No - press enter \\n'))\n",
    "    pathway_names = pathway_names_input.split(' ')\n",
    "\n",
    "\n",
    "# inform about requested IDs\n",
    "\n",
    "print('\\n List of Reactome IDs saved to be examined later:') \n",
    "for i in range(len(reactome_ids)):                                           \n",
    "    if pathway_names[0] != '':\n",
    "        print(reactome_ids[i], ' (', pathway_names[i], ')', sep='')\n",
    "    else:\n",
    "        print(reactome_ids[i])\n",
    "\n",
    "print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Download the symbols from Reactome and check if they are HGNC-approved\n",
    "**Description of the script:**\n",
    "1. Gene symbols from Reactome are downloaded (using Reactome IDs requested earlier)\n",
    "2. The symbols are checked to be HGNC-approved\n",
    "3. List stress_molecules_list is created; containing HGNC-approved symbols, each only once\n",
    "4. Dictionary 'gene_in_multiple_paths_check' is generated to identify genes that are involved in multiple Reactome pathways.\n",
    "\n",
    "Results:\n",
    "1. stress_molecules_list\n",
    "    - stress_molecules.csv\n",
    "2. gene_in_multiple_paths_check \n",
    "    - gene_in_multiple_paths_check.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene symbols of pathway: hsr (ID: R-HSA-3371556) are being downloaded and checked...\n",
      "Gene symbols of pathway: autophagy (ID: R-HSA-9612973) are being downloaded and checked...\n",
      "Gene symbols of pathway: starvation (ID: R-HSA-9711097) are being downloaded and checked...\n",
      "Gene symbols of pathway: upr (ID: R-HSA-381119) are being downloaded and checked...\n",
      "Gene symbols of pathway: dna-repair (ID: R-HSA-73894) are being downloaded and checked...\n",
      "\n",
      " Done.\n"
     ]
    }
   ],
   "source": [
    "for k in range(len(reactome_ids)): # for every Reactome ID requested:\n",
    "    \n",
    "    if pathway_names[0] != '':    \n",
    "        print('Gene symbols of pathway: {0} (ID: {1}) are being downloaded and checked...'.format(pathway_names[k], reactome_ids[k]))\n",
    "    else: \n",
    "        print('Gene symbols of pathway with ID: {} are being downloaded and checked...'.format(reactome_ids[k]))\n",
    "    \n",
    "    \n",
    "    reactome_id = reactome_ids[k]\n",
    "    \n",
    "    working_symbols_list = []        # list of gene symbols of current Reactome ID to be HGNC-checked\n",
    "    \n",
    "    HGNC_approved_symbols_list = []\n",
    "    \n",
    "    \n",
    "    \n",
    "    # download a list of molecules using Reactome API \n",
    "    \n",
    "    headers = {'Accept': 'application/json'}\n",
    "\n",
    "    uri = 'https://reactome.org/ContentService/data/participants'\n",
    "    path = '/{}'.format(reactome_id)\n",
    "\n",
    "    target = urlparse(uri+path)\n",
    "    method = 'GET'\n",
    "    body = ''\n",
    "\n",
    "    h = http.Http()\n",
    "\n",
    "    response, content = h.request(\n",
    "        target.geturl(),\n",
    "        method,\n",
    "        body,\n",
    "        headers)\n",
    "\n",
    "\n",
    "    if response['status'] == '200':    # assume that content is a json reply\n",
    "        data = json.loads(content)     # parse content with the json module \n",
    "\n",
    "\n",
    "        for i in range(len(data)):\n",
    "            for j in range(len(data[i]['refEntities'])):\n",
    "                if 'UniProt' in data[i]['refEntities'][j]['displayName']:         # choose only UniProt outcomes; i.e. gene symbols\n",
    "                    symbol = data[i]['refEntities'][j]['displayName'].split()[1]\n",
    "                    if symbol not in working_symbols_list:                        # check that every symbol is only once\n",
    "                        working_symbols_list.append(symbol)\n",
    "\n",
    "    else:\n",
    "        print( 'Error detected: ' + response['status'])\n",
    "           \n",
    "    \n",
    "    \n",
    "    # with using HGNC API find out if symbols are HGNC-approved\n",
    "\n",
    "    if len(working_symbols_list) != 0:                 # if list is not empty\n",
    "        for symbol in working_symbols_list:            # check every symbol in list\n",
    "            \n",
    "            # set the url\n",
    "            headers = {'Accept': 'application/json'}\n",
    "\n",
    "            uri = 'https://rest.genenames.org'\n",
    "            path = '/search/symbol/{}'.format(symbol)\n",
    "\n",
    "            target = urlparse(uri+path)\n",
    "            method = 'GET'\n",
    "            body = ''\n",
    "\n",
    "            h = http.Http()\n",
    "\n",
    "            response, content = h.request(\n",
    "                target.geturl(),\n",
    "                method,\n",
    "                body,\n",
    "                headers)\n",
    "\n",
    "            if response['status'] == '200':            # assume that content is a json reply\n",
    "                data = json.loads(content)             # parse content with the json module \n",
    "                check = data['response']['numFound']   # check can be 0 (not HGNC-approved) or 1 (approved)\n",
    "\n",
    "                if (check == 0) and (symbol not in HGNC_not):\n",
    "                    HGNC_not.append(symbol)\n",
    "                elif (check == 1) and (symbol not in HGNC_approved_symbols_list):\n",
    "                    HGNC_approved_symbols_list.append(symbol)\n",
    "\n",
    "            else:\n",
    "                print( 'Error detected: ' + response['status'])\n",
    "        \n",
    "        reactome_ids_symbols_dictionary[reactome_id] = HGNC_approved_symbols_list      # 'reactome_id_1: [gene_1, gene2, ...]' is added to dictionary\n",
    "        \n",
    "\n",
    "    else: \n",
    "        print('Error: there is no data for {}.'.format(reactome_id))\n",
    "    \n",
    "    for gene in HGNC_approved_symbols_list:\n",
    "        if gene not in stress_molecules_list: \n",
    "            stress_molecules_list.append(gene)  # HGNC-approved symbols added to list of stress molecules (--> stress_molecules.csv)\n",
    "        \n",
    "        if gene not in gene_in_multiple_paths_check.keys():\n",
    "            gene_in_multiple_paths_check.update({gene: [reactome_id]})\n",
    "        else:\n",
    "            gene_in_multiple_paths_check[gene].append(reactome_id)\n",
    "        \n",
    "\n",
    "\n",
    "# print symbols that are not HGNC-approved\n",
    "\n",
    "if len(HGNC_not) != 0:    \n",
    "    print('\\n Symbols not HGNC-approved:', sep = ' ')\n",
    "    for gene in HGNC_not:\n",
    "        print(gene, sep = ' ')\n",
    "\n",
    "        \n",
    "\n",
    "df_stress_molecules = pd.DataFrame(stress_molecules_list, columns = ['symbol'])\n",
    "    \n",
    "multiple_paths_df = pd.DataFrame.from_dict(gene_in_multiple_paths_check, orient = 'index')\n",
    "\n",
    "\n",
    "print('\\n Done.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results: List of stress molecules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EP300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CREBBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HSPA1L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HSPA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HSPA8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>746</th>\n",
       "      <td>MSH2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>747</th>\n",
       "      <td>MSH6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>MLH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>PMS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750</th>\n",
       "      <td>MSH3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>751 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     symbol\n",
       "0     EP300\n",
       "1    CREBBP\n",
       "2    HSPA1L\n",
       "3     HSPA2\n",
       "4     HSPA8\n",
       "..      ...\n",
       "746    MSH2\n",
       "747    MSH6\n",
       "748    MLH1\n",
       "749    PMS2\n",
       "750    MSH3\n",
       "\n",
       "[751 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stress_molecules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Results: Multiple paths check\n",
    "To identify genes that are involved in multiple Reactome pathways."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EP300</th>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-73894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CREBBP</th>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPA1L</th>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPA2</th>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPA8</th>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-9612973</td>\n",
       "      <td>R-HSA-3371556</td>\n",
       "      <td>R-HSA-9612973</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSH2</th>\n",
       "      <td>R-HSA-73894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSH6</th>\n",
       "      <td>R-HSA-73894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MLH1</th>\n",
       "      <td>R-HSA-73894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMS2</th>\n",
       "      <td>R-HSA-73894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSH3</th>\n",
       "      <td>R-HSA-73894</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>751 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0              1              2              3  \\\n",
       "EP300   R-HSA-3371556  R-HSA-3371556  R-HSA-3371556    R-HSA-73894   \n",
       "CREBBP  R-HSA-3371556  R-HSA-3371556  R-HSA-3371556           None   \n",
       "HSPA1L  R-HSA-3371556  R-HSA-3371556  R-HSA-3371556           None   \n",
       "HSPA2   R-HSA-3371556  R-HSA-3371556  R-HSA-3371556           None   \n",
       "HSPA8   R-HSA-3371556  R-HSA-3371556  R-HSA-9612973  R-HSA-3371556   \n",
       "...               ...            ...            ...            ...   \n",
       "MSH2      R-HSA-73894           None           None           None   \n",
       "MSH6      R-HSA-73894           None           None           None   \n",
       "MLH1      R-HSA-73894           None           None           None   \n",
       "PMS2      R-HSA-73894           None           None           None   \n",
       "MSH3      R-HSA-73894           None           None           None   \n",
       "\n",
       "                    4     5  \n",
       "EP300            None  None  \n",
       "CREBBP           None  None  \n",
       "HSPA1L           None  None  \n",
       "HSPA2            None  None  \n",
       "HSPA8   R-HSA-9612973  None  \n",
       "...               ...   ...  \n",
       "MSH2             None  None  \n",
       "MSH6             None  None  \n",
       "MLH1             None  None  \n",
       "PMS2             None  None  \n",
       "MSH3             None  None  \n",
       "\n",
       "[751 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiple_paths_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2 Get orthologs for every \"stress molecule\"; from NCBI\n",
    "**Description of the script:** \n",
    "1. Download and save FASTA files using NCBI Datasets\n",
    "    - FASTA file (orthologs) for every stress molecule using NCBI Datasets (datasets.exe) is downloaded and saved\n",
    "2. Parse FASTA files\n",
    "    - FASTA files are parsed and saved to: \n",
    "        - orthologs_list with all information from FASTA file saved\n",
    "        - orthologs_count for analysis of count of sequences per gene id\n",
    "        - genes_organisms_dictionary for genes-organisms check\n",
    "        - all_orthologs_refseqs_sorted for selection of one sequence per gene id\n",
    "3. Genes-organisms check\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Download and save FASTA files using NCBI Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FASTA files for every stress molecule containing data with orthologous sequences are being downloaded and saved... \n",
      "\n",
      "Error: a file for HSPA7 was not downloaded. \n",
      "\n",
      "Error: a file for GABARAPL3 was not downloaded. \n",
      "\n",
      "Error: a file for H2BC12L was not downloaded. \n",
      "\n",
      "Error: a file for H2BC13 was not downloaded. \n",
      "\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "print('FASTA files for every stress molecule containing data with orthologous sequences are being downloaded and saved... \\n') \n",
    "\n",
    "for stress_molecule in stress_molecules_list:\n",
    "    os.system('datasets download ortholog symbol {1} --taxon-filter mammals --exclude-rna --exclude-gene --filename {0}\\\\NCBI_Datasets_download\\\\{1}.zip'.format(win_path, stress_molecule))\n",
    "    os.system('cd {0}\\\\NCBI_Datasets_download\\\\ & tar -xf {0}\\\\NCBI_Datasets_download\\\\{1}.zip'.format(win_path, stress_molecule))\n",
    "    os.system('move {0}\\\\NCBI_Datasets_download\\\\ncbi_dataset\\\\data\\\\protein.faa {0}\\\\NCBI_Datasets_download'.format(win_path))\n",
    "    os.system('ren {0}\\\\NCBI_Datasets_download\\\\protein.faa {1}.faa'.format(win_path, stress_molecule))\n",
    "os.system('mkdir {0}\\\\NCBI_Datasets_download\\\\hej'.format(win_path))\n",
    "os.system('move {0}\\\\NCBI_Datasets_download\\\\README.md {0}\\\\NCBI_Datasets_download\\\\hej'.format(win_path))\n",
    "os.system('move {0}\\\\NCBI_Datasets_download\\\\*.zip {0}\\\\NCBI_Datasets_download\\\\hej'.format(win_path))\n",
    "os.system('rmdir /s /q {0}\\\\NCBI_Datasets_download\\\\hej'.format(win_path))\n",
    "os.system('rmdir /s /q {0}\\\\NCBI_Datasets_download\\\\ncbi_dataset'.format(win_path))\n",
    "\n",
    "\n",
    "files = os.listdir(r'{0}\\NCBI_Datasets_download'.format(main_path))\n",
    "if '.ipynb_checkpoints' in files: files.remove('.ipynb_checkpoints')\n",
    "\n",
    "\n",
    "for stress_molecule in stress_molecules_list:\n",
    "    if str(stress_molecule) + '.faa' not in files:\n",
    "        print('Error: a file for', stress_molecule, 'was not downloaded. \\n')\n",
    "        stress_molecules_list.remove(stress_molecule)\n",
    "\n",
    "print('Done.')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Parse FASTA files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Variables and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENERAL VARIABLES\n",
    "gene_ids = [] \n",
    "# gene ids of all orthologs (of all stress molecules together)\n",
    "\n",
    "orthologs_list = [] \n",
    "# list of all orthologous sequences and information about them\n",
    "# [{human_gene_name: X, gene id: X, gene name: X, organism name; X, accession_number (that is refseq number): X, protein sequence: X}, {...}]\n",
    "# -> orthologs.csv\n",
    "\n",
    "\n",
    "# TO SEE HOW MANY REFERENCE SEQUENCES THERE ARE FOR ONE GENE ID\n",
    "orthologs_count_list = []\n",
    "# list of genes and count of its sequences (refseq) downloaded from NCBI \n",
    "# [{human_gene_name: x, gene_id: x, seq_count = x, most_repres_seq_count = x, unique_seq_count = x}, {...}, ...]\n",
    "\n",
    "already_been_there = []\n",
    "# working list\n",
    "\n",
    "\n",
    "# FOR GENES-ORGANISMS CHECK\n",
    "genes_organisms_dictionary = {}\n",
    "\n",
    "\n",
    "\n",
    "# FOR SELECTION OF ONE SEQUENCE PER GENE ID\n",
    "all_orthologs_refseqs_sorted = {}\n",
    "# {gene_id1: \n",
    "#     {RefSeq1: \n",
    "#          {protein_sequence: X, human_gene_name: X, gene_name: X, organism: X}, \n",
    "#      RefSeq2: {...}\n",
    "#     }, \n",
    "#  gene_id2: \n",
    "#      {RefSeq3: {}\n",
    "#     }, \n",
    "#  ...}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def most_frequent_count(List):\n",
    "    occurence_count = Counter(List)\n",
    "    return occurence_count.most_common(1)[0][1]\n",
    "\n",
    "def most_frequent_name(List):\n",
    "    occurence_count = Counter(List)\n",
    "    return occurence_count.most_common(1)[0][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Script: Parse FASTA files and save the information into lists and dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# parse all FASTA files downloaded \n",
    "# every file is named as the gene which orthologs seqs it contains of\n",
    "\n",
    "files = os.listdir(r'{0}\\NCBI_Datasets_download'.format(main_path))\n",
    "if '.ipynb_checkpoints'in files: files.remove('.ipynb_checkpoints')\n",
    "\n",
    "for file in files:\n",
    "    \n",
    "    gene_name = file.replace('.faa', '')\n",
    "\n",
    "    orthologs_count_dictionary = {}   # working dictionary; will be used for \"seqs count check\" \n",
    "                                      # {gene_id1: [seq1, seq2, seq3], gene_id2: [seq1, seq2], ...}\n",
    "    \n",
    "    organism_gene_id_dict = {}        # working dictionary; will be used for genes-organisms check\n",
    "        \n",
    "    with open(r'{0}\\NCBI_Datasets_download\\{1}'.format(main_path, file)) as fasta_file:\n",
    "        parser = fastaparser.Reader(fasta_file)\n",
    "        \n",
    "        for seq in parser: # for every record (ortholog) of FASTA file        \n",
    "            \n",
    "            # split the FASTA record \n",
    "            working_list = seq.description.split(\" [\")     \n",
    "            gene_id = (re.findall('\\\\d+', working_list[2])[0])\n",
    "            gene_ids.append(gene_id)\n",
    "\n",
    "\n",
    "            \n",
    "            ### for orthologs.csv ###\n",
    "            # add seq (ortholog) atributes to orthologs_list\n",
    "            \n",
    "            ortholog_dictionary = {}   # {human_gene_name: X, gene id: X, gene name: X, organism name; X, accession number: X, protein sequence: X}\n",
    "            \n",
    "            ortholog_dictionary['human_gene_name'] = gene_name                              # human gene name added\n",
    "            ortholog_dictionary['gene_name'] = working_list[0]                              # gene name added\n",
    "            organism_name = working_list[1][9:].replace(\"]\", \"\")\n",
    "            organism_name = organism_name.replace(\" \", \"_\")\n",
    "            ortholog_dictionary['organism'] = organism_name                                 # organism name added\n",
    "            ortholog_dictionary['gene_id'] = gene_id                                        # gene ID added\n",
    "            ortholog_dictionary['accession_number'] = seq.id                                # accession number added \n",
    "            ortholog_dictionary['protein_sequence'] = seq.sequence_as_string()              # protein sequence with its accession number added\n",
    "\n",
    "            orthologs_list.append(ortholog_dictionary)\n",
    "            # orthologs_list = [{human_gene_name: X, gene id: X, gene name: X, organism name; X, accession number: X, protein sequence: X}, {...}]\n",
    "            \n",
    "            \n",
    "\n",
    "            ### for orthologs_count.csv and final_orthologs_seqs.csv ###     \n",
    "            # add sequence to orthologs_count_dictionary {gene_id1: [seq1, seq2, seq3], gene_id2: [seq1, seq2], ...}\n",
    "            # add RefSeq number and sequence to all_orthologs_refseqs_sorted dictionary {gene_id1: {RefSeq1: seq1, RefSeq2: seq2, ...}, gene_id2: {}, ...}\n",
    "            \n",
    "            if gene_id not in already_been_there: \n",
    "                first_sequence = seq.sequence_as_string()\n",
    "                orthologs_count_dictionary[gene_id] = [first_sequence]\n",
    "                all_orthologs_refseqs_sorted[gene_id] = {}\n",
    "                all_orthologs_refseqs_sorted[gene_id].update({seq.id: {'protein_sequence': seq.sequence_as_string(), 'human_gene_name': gene_name, 'gene_name': working_list[0], 'organism': working_list[1][9:].replace(\"]\", \"\")}})\n",
    "                \n",
    "            else:\n",
    "                another_sequence = seq.sequence_as_string()\n",
    "                orthologs_count_dictionary[gene_id].append(another_sequence)\n",
    "                all_orthologs_refseqs_sorted[gene_id].update({seq.id: {'protein_sequence': seq.sequence_as_string(), 'human_gene_name': gene_name, 'gene_name': working_list[0], 'organism': working_list[1][9:].replace(\"]\", \"\")}})\n",
    "\n",
    "            already_been_there.append(gene_id)\n",
    "            \n",
    "            \n",
    "            \n",
    "            ### for genes_organisms_check.csv ###\n",
    "            organism_gene_id_dict[organism_name] = gene_id      \n",
    "\n",
    "            \n",
    "        genes_organisms_dictionary.update({gene_name : organism_gene_id_dict})\n",
    "        # {gene_name1: {organism1: gene_id1, organism2: gene_id2, ...}, gene_name2: {}, ...}\n",
    "            \n",
    "            \n",
    "    \n",
    "    ### for orthologs_count.csv ###     \n",
    "    # create dictionary with gene ids and their protein sequence counts (and the most represented sequence)\n",
    "    \n",
    "    for key in orthologs_count_dictionary.keys():\n",
    "        ortholog = {}\n",
    "        ortholog['human_gene_name'] = file.strip('.faa')\n",
    "        ortholog['gene_id'] = str(key)\n",
    "        ortholog['seq_count'] = len(orthologs_count_dictionary[key])\n",
    "        ortholog['most_repres_seq_count'] = most_frequent_count(orthologs_count_dictionary[key])\n",
    "        \n",
    "        counter_keys = Counter(orthologs_count_dictionary[key])\n",
    "        ortholog['unique_seq_count'] = len(counter_keys.keys())\n",
    "        \n",
    "        orthologs_count_list.append(ortholog)\n",
    "        # orthologs_count_list = [{human_gene_name: 'xxx.faa', gene_id: x, seq_count = x, most_repres_seq_count = x, unique_seq_count = x}, {...}, ...]\n",
    "\n",
    "\n",
    "df_orthologs_list = pd.DataFrame.from_dict(orthologs_list)\n",
    "df_ortholog_count_list = pd.DataFrame.from_dict(orthologs_count_list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Results: List of all orthologs downloaded "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_gene_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>Equus_caballus</td>\n",
       "      <td>100063811</td>\n",
       "      <td>XP_001494960.1</td>\n",
       "      <td>MCSLGLFPPPPPRGQVTLYEHNNELVTGNSYESPPPDFRGQWINLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>Equus_caballus</td>\n",
       "      <td>100063811</td>\n",
       "      <td>XP_023499419.1</td>\n",
       "      <td>MRVAPSFSPVHSTGRDVGLFGVLNEIANSEEEVFEWVKTASSWALA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>Sus_scrofa</td>\n",
       "      <td>100154333</td>\n",
       "      <td>XP_003355451.1</td>\n",
       "      <td>MCSLGLFPPPPPRGQVTLYEHNNELVTGSSYESPPPDFRGQWINLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>Oryctolagus_cuniculus</td>\n",
       "      <td>100356621</td>\n",
       "      <td>XP_002711061.1</td>\n",
       "      <td>MCSLGLFPPPPPMGQVTLYEHNNELVTGSSYESPPPDFRGQWINLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>Oryctolagus_cuniculus</td>\n",
       "      <td>100356621</td>\n",
       "      <td>XP_008254733.1</td>\n",
       "      <td>MARGLPSSITGSKCGKDASTRVFEWVKTASSWALALCRWASSLHGS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359400</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>Zfp830</td>\n",
       "      <td>Rattus_norvegicus</td>\n",
       "      <td>497967</td>\n",
       "      <td>NP_001032437.1</td>\n",
       "      <td>MASSTSARTPAGKRVVNQEELRRLMKEKQRLSTNRKRIESPFAKYN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359401</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>ZNF830</td>\n",
       "      <td>Bos_taurus</td>\n",
       "      <td>539497</td>\n",
       "      <td>NP_001076081.1</td>\n",
       "      <td>MASSASARPPAGKRVVNQDELRRLMKEKQRLSTNRKRIESPFAKYN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359402</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>Zfp830</td>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>66983</td>\n",
       "      <td>NP_080160.2</td>\n",
       "      <td>MASSTSTRTPAGKRVVNQEELRRLMREKQRLSTNRKRIESPFAKYN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359403</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>ZNF830</td>\n",
       "      <td>Macaca_mulatta</td>\n",
       "      <td>715130</td>\n",
       "      <td>XP_001113549.3</td>\n",
       "      <td>MPRRTETRIALGLVAKMASSASARTPAGKRVVNQEELRRLMKEKQR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359404</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>ZNF830</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>91603</td>\n",
       "      <td>NP_443089.3</td>\n",
       "      <td>MASSASARTPAGKRVINQEELRRLMKEKQRLSTSRKRIESPFAKYN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>359405 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       human_gene_name gene_name               organism    gene_id  \\\n",
       "0                 AAAS      AAAS         Equus_caballus  100063811   \n",
       "1                 AAAS      AAAS         Equus_caballus  100063811   \n",
       "2                 AAAS      AAAS             Sus_scrofa  100154333   \n",
       "3                 AAAS      AAAS  Oryctolagus_cuniculus  100356621   \n",
       "4                 AAAS      AAAS  Oryctolagus_cuniculus  100356621   \n",
       "...                ...       ...                    ...        ...   \n",
       "359400          ZNF830    Zfp830      Rattus_norvegicus     497967   \n",
       "359401          ZNF830    ZNF830             Bos_taurus     539497   \n",
       "359402          ZNF830    Zfp830           Mus_musculus      66983   \n",
       "359403          ZNF830    ZNF830         Macaca_mulatta     715130   \n",
       "359404          ZNF830    ZNF830           Homo_sapiens      91603   \n",
       "\n",
       "       accession_number                                   protein_sequence  \n",
       "0        XP_001494960.1  MCSLGLFPPPPPRGQVTLYEHNNELVTGNSYESPPPDFRGQWINLP...  \n",
       "1        XP_023499419.1  MRVAPSFSPVHSTGRDVGLFGVLNEIANSEEEVFEWVKTASSWALA...  \n",
       "2        XP_003355451.1  MCSLGLFPPPPPRGQVTLYEHNNELVTGSSYESPPPDFRGQWINLP...  \n",
       "3        XP_002711061.1  MCSLGLFPPPPPMGQVTLYEHNNELVTGSSYESPPPDFRGQWINLP...  \n",
       "4        XP_008254733.1  MARGLPSSITGSKCGKDASTRVFEWVKTASSWALALCRWASSLHGS...  \n",
       "...                 ...                                                ...  \n",
       "359400   NP_001032437.1  MASSTSARTPAGKRVVNQEELRRLMKEKQRLSTNRKRIESPFAKYN...  \n",
       "359401   NP_001076081.1  MASSASARPPAGKRVVNQDELRRLMKEKQRLSTNRKRIESPFAKYN...  \n",
       "359402      NP_080160.2  MASSTSTRTPAGKRVVNQEELRRLMREKQRLSTNRKRIESPFAKYN...  \n",
       "359403   XP_001113549.3  MPRRTETRIALGLVAKMASSASARTPAGKRVVNQEELRRLMKEKQR...  \n",
       "359404      NP_443089.3  MASSASARTPAGKRVINQEELRRLMKEKQRLSTSRKRIESPFAKYN...  \n",
       "\n",
       "[359405 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_orthologs_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Results: Count of reference sequences of every orthologous gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seq_count</th>\n",
       "      <th>most_repres_seq_count</th>\n",
       "      <th>unique_seq_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>100063811</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>100154333</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>100356621</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>100405154</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>100444879</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138826</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>497967</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138827</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>539497</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138828</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>66983</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138829</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>715130</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138830</th>\n",
       "      <td>ZNF830</td>\n",
       "      <td>91603</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138831 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       human_gene_name    gene_id  seq_count  most_repres_seq_count  \\\n",
       "0                 AAAS  100063811          2                      1   \n",
       "1                 AAAS  100154333          1                      1   \n",
       "2                 AAAS  100356621          4                      3   \n",
       "3                 AAAS  100405154          2                      1   \n",
       "4                 AAAS  100444879          4                      1   \n",
       "...                ...        ...        ...                    ...   \n",
       "138826          ZNF830     497967          1                      1   \n",
       "138827          ZNF830     539497          1                      1   \n",
       "138828          ZNF830      66983          1                      1   \n",
       "138829          ZNF830     715130          1                      1   \n",
       "138830          ZNF830      91603          1                      1   \n",
       "\n",
       "        unique_seq_count  \n",
       "0                      2  \n",
       "1                      1  \n",
       "2                      2  \n",
       "3                      2  \n",
       "4                      4  \n",
       "...                  ...  \n",
       "138826                 1  \n",
       "138827                 1  \n",
       "138828                 1  \n",
       "138829                 1  \n",
       "138830                 1  \n",
       "\n",
       "[138831 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ortholog_count_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Genes-organisms check\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There is not enough data for genes:\n",
      "CCL2\n",
      "CENPS\n",
      "CHMP3\n",
      "ERCC5\n",
      "GTF2H2\n",
      "H2AB1\n",
      "H2AC14\n",
      "H2AC18\n",
      "H2AC20\n",
      "H2AC4\n",
      "H2AC6\n",
      "H2AC7\n",
      "H2BC1\n",
      "H2BC11\n",
      "H2BC12\n",
      "H2BC14\n",
      "H2BC15\n",
      "H2BC17\n",
      "H2BC21\n",
      "H2BC26\n",
      "H2BC3\n",
      "H2BC4\n",
      "H2BC5\n",
      "H2BC9\n",
      "H3-4\n",
      "H4C1\n",
      "HBB\n",
      "HSPA1A\n",
      "HSPA1B\n",
      "HSPA1L\n",
      "HSPA6\n",
      "ISY1\n",
      "MAP1LC3C\n",
      "PLA2G4B\n",
      "POM121\n",
      "POM121C\n",
      "RANBP2\n",
      "RHNO1\n",
      "RNASE1\n",
      "RPL10L\n",
      "RPL17\n",
      "RPL36A\n",
      "RPL39L\n",
      "RPL41\n",
      "RPS10\n",
      "RPS4Y1\n",
      "RPS4Y2\n",
      "RRAGB\n",
      "SEM1\n",
      "SRPRB\n",
      "SULT1A3\n",
      "TUBA1A\n",
      "TUBA1B\n",
      "TUBA1C\n",
      "TUBA3C\n",
      "TUBA3D\n",
      "TUBA3E\n",
      "TUBA4A\n",
      "TUBA4B\n",
      "TUBB2A\n",
      "TUBB2B\n",
      "TUBB3\n",
      "TUBB8\n",
      "TUBB8B\n",
      "UBB\n",
      "UBC\n",
      "There is not enough data for organisms:\n",
      "Elephantulus_edwardii\n",
      "\n",
      " They are not included in future analysis.\n"
     ]
    }
   ],
   "source": [
    "# create DataFrame from genes_organisms_dictionary\n",
    "\n",
    "df_genes_organisms = pd.DataFrame(list(genes_organisms_dictionary.values()), index=list(genes_organisms_dictionary.keys()))\n",
    "\n",
    "# variables\n",
    "\n",
    "genes_count = len(genes_organisms_dictionary.keys())\n",
    "organisms_count = 0\n",
    "count_organisms_seq_ratio = {}\n",
    "count_genes_seq_ratio = {}\n",
    "\n",
    "# for which organisms there is enough data? \n",
    "# i.e. ratio of nonexisting vs existing ortholog seqs for every organism\n",
    "for column in df_genes_organisms:\n",
    "    organisms_count += 1\n",
    "    count_organisms_seq_ratio.update({column: df_genes_organisms[column].isnull().sum()/genes_count})\n",
    "\n",
    "# for which genes there is enough data? \n",
    "# i.e. ratio of nonexisting vs existing seqs for every gene\n",
    "for key in genes_organisms_dictionary.keys():\n",
    "    nan_count = df_genes_organisms.loc[[key]].isna().sum().sum()\n",
    "    count_genes_seq_ratio.update({key: nan_count/organisms_count})\n",
    "\n",
    "\n",
    "\n",
    "# pop an organism or a gene from the DataFrame if there is not enough data (seqs) for them \n",
    "\n",
    "organisms_not_enough = []\n",
    "genes_not_enough = []\n",
    "gene_ids_to_pop = [] # list of gene ids to pop from all_orthologs_refseq_sorted (aka list for future analysis)\n",
    "\n",
    "for key in count_organisms_seq_ratio.keys():\n",
    "    if count_organisms_seq_ratio[key] > 0.2:\n",
    "        organisms_not_enough.append(key)\n",
    "        \n",
    "\n",
    "# Remove the organism with too litle data from the analysis\n",
    "for organism in organisms_not_enough:\n",
    "    gene_ids_to_pop += df_genes_organisms[organism].to_list()\n",
    "    df_genes_organisms.pop(organism)\n",
    "    \n",
    "    \n",
    "\n",
    "for key in count_genes_seq_ratio.keys():\n",
    "    if count_genes_seq_ratio[key] > 0.2:\n",
    "        genes_not_enough.append(key)\n",
    "        \n",
    "        \n",
    "# Remove the gene with too litle data from the analysis\n",
    "for gene in genes_not_enough:\n",
    "    gene_ids_to_pop += df_genes_organisms.loc[gene].tolist()\n",
    "    df_genes_organisms.drop(gene, inplace=True)\n",
    "\n",
    "\n",
    "\n",
    "# print genes and organisms which enough data do not exist for\n",
    "\n",
    "if len(genes_not_enough) != 0:\n",
    "    print('There is not enough data for genes:')\n",
    "    for gene in genes_not_enough:\n",
    "        print(gene)\n",
    "\n",
    "if len(organisms_not_enough) != 0:\n",
    "    print('There is not enough data for organisms:')\n",
    "    for organism in organisms_not_enough:\n",
    "        print(organism)\n",
    "        \n",
    "print('\\n They are not included in future analysis.')\n",
    "\n",
    "\n",
    "\n",
    "# remove organisms and genes with not enough data from all_orthologs_refseqs_sorted\n",
    "gene_ids_to_pop = [b for b in gene_ids_to_pop if isinstance(b, str)] # remove all nan\n",
    "\n",
    "\n",
    "for geneid in gene_ids_to_pop:\n",
    "    del all_orthologs_refseqs_sorted[geneid]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Results: Final table of genes, organism and corresponding gene ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Equus_caballus</th>\n",
       "      <th>Sus_scrofa</th>\n",
       "      <th>Oryctolagus_cuniculus</th>\n",
       "      <th>Callithrix_jacchus</th>\n",
       "      <th>Pongo_abelii</th>\n",
       "      <th>Ailuropoda_melanoleuca</th>\n",
       "      <th>Nomascus_leucogenys</th>\n",
       "      <th>Loxodonta_africana</th>\n",
       "      <th>Cavia_porcellus</th>\n",
       "      <th>Cricetulus_griseus</th>\n",
       "      <th>...</th>\n",
       "      <th>Mus_musculus</th>\n",
       "      <th>Rattus_norvegicus</th>\n",
       "      <th>Pan_troglodytes</th>\n",
       "      <th>Bos_taurus</th>\n",
       "      <th>Canis_lupus_familiaris</th>\n",
       "      <th>Macaca_mulatta</th>\n",
       "      <th>Homo_sapiens</th>\n",
       "      <th>Monodelphis_domestica</th>\n",
       "      <th>Sturnira_hondurensis</th>\n",
       "      <th>Microtus_oregoni</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>100063811</td>\n",
       "      <td>100154333</td>\n",
       "      <td>100356621</td>\n",
       "      <td>100405154</td>\n",
       "      <td>100444879</td>\n",
       "      <td>100468100</td>\n",
       "      <td>100586295</td>\n",
       "      <td>100662257</td>\n",
       "      <td>100727054</td>\n",
       "      <td>100767976</td>\n",
       "      <td>...</td>\n",
       "      <td>223921</td>\n",
       "      <td>300259</td>\n",
       "      <td>467003</td>\n",
       "      <td>506561</td>\n",
       "      <td>607867</td>\n",
       "      <td>719095</td>\n",
       "      <td>8086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>100069715</td>\n",
       "      <td>100524544</td>\n",
       "      <td>100345909</td>\n",
       "      <td>100398414</td>\n",
       "      <td>100453216</td>\n",
       "      <td>100466468</td>\n",
       "      <td>100601002</td>\n",
       "      <td>100677152</td>\n",
       "      <td>100720976</td>\n",
       "      <td>100768098</td>\n",
       "      <td>...</td>\n",
       "      <td>11350</td>\n",
       "      <td>311860</td>\n",
       "      <td>107976801</td>\n",
       "      <td>540876</td>\n",
       "      <td>491292</td>\n",
       "      <td>722449</td>\n",
       "      <td>25</td>\n",
       "      <td>100619549</td>\n",
       "      <td>118980343</td>\n",
       "      <td>121462621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABRAXAS1</th>\n",
       "      <td>100061483</td>\n",
       "      <td>100524450</td>\n",
       "      <td>100350582</td>\n",
       "      <td>100401953</td>\n",
       "      <td>100449014</td>\n",
       "      <td>100463877</td>\n",
       "      <td>100597649</td>\n",
       "      <td>100675954</td>\n",
       "      <td>100721329</td>\n",
       "      <td>100766855</td>\n",
       "      <td>...</td>\n",
       "      <td>70681</td>\n",
       "      <td>289468</td>\n",
       "      <td>461205</td>\n",
       "      <td>504796</td>\n",
       "      <td>478459</td>\n",
       "      <td>713263</td>\n",
       "      <td>84142</td>\n",
       "      <td>100017569</td>\n",
       "      <td>118979544</td>\n",
       "      <td>121440920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACADVL</th>\n",
       "      <td>100061583</td>\n",
       "      <td>100520636</td>\n",
       "      <td>100346654</td>\n",
       "      <td>100399197</td>\n",
       "      <td>100454381</td>\n",
       "      <td>100472785</td>\n",
       "      <td>100585921</td>\n",
       "      <td>100657276</td>\n",
       "      <td>100713909</td>\n",
       "      <td>100767527</td>\n",
       "      <td>...</td>\n",
       "      <td>11370</td>\n",
       "      <td>25363</td>\n",
       "      <td>455237</td>\n",
       "      <td>282130</td>\n",
       "      <td>489463</td>\n",
       "      <td>714326</td>\n",
       "      <td>37</td>\n",
       "      <td>100016619</td>\n",
       "      <td>118974685</td>\n",
       "      <td>121456541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACD</th>\n",
       "      <td>100066184</td>\n",
       "      <td>100623181</td>\n",
       "      <td>100346875</td>\n",
       "      <td>100405339</td>\n",
       "      <td>100447506</td>\n",
       "      <td>100474125</td>\n",
       "      <td>100586139</td>\n",
       "      <td>100661748</td>\n",
       "      <td>100730199</td>\n",
       "      <td>100763518</td>\n",
       "      <td>...</td>\n",
       "      <td>497652</td>\n",
       "      <td>307798</td>\n",
       "      <td>454170</td>\n",
       "      <td>510353</td>\n",
       "      <td>487931</td>\n",
       "      <td>699230</td>\n",
       "      <td>65057</td>\n",
       "      <td>103105155</td>\n",
       "      <td>118983853</td>\n",
       "      <td>121447226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YIF1A</th>\n",
       "      <td>100052354</td>\n",
       "      <td>100521064</td>\n",
       "      <td>127489740</td>\n",
       "      <td>100394171</td>\n",
       "      <td>100443779</td>\n",
       "      <td>100482492</td>\n",
       "      <td>100597699</td>\n",
       "      <td>100656055</td>\n",
       "      <td>100727936</td>\n",
       "      <td>100763610</td>\n",
       "      <td>...</td>\n",
       "      <td>68090</td>\n",
       "      <td>171441</td>\n",
       "      <td>746258</td>\n",
       "      <td>506336</td>\n",
       "      <td>476014</td>\n",
       "      <td>714486</td>\n",
       "      <td>10897</td>\n",
       "      <td>100029722</td>\n",
       "      <td>118991769</td>\n",
       "      <td>121454377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YWHAE</th>\n",
       "      <td>100059962</td>\n",
       "      <td>100620600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100400280</td>\n",
       "      <td>100455220</td>\n",
       "      <td>100484851</td>\n",
       "      <td>100600918</td>\n",
       "      <td>100658987</td>\n",
       "      <td>100714396</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>22627</td>\n",
       "      <td>29753</td>\n",
       "      <td>454409</td>\n",
       "      <td>282125</td>\n",
       "      <td>480645</td>\n",
       "      <td>720156</td>\n",
       "      <td>7531</td>\n",
       "      <td>100012276</td>\n",
       "      <td>119003777</td>\n",
       "      <td>121456736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YY1</th>\n",
       "      <td>100630389</td>\n",
       "      <td>102162035</td>\n",
       "      <td>103346032</td>\n",
       "      <td>100411814</td>\n",
       "      <td>100447742</td>\n",
       "      <td>100474192</td>\n",
       "      <td>100583435</td>\n",
       "      <td>100673864</td>\n",
       "      <td>100730873</td>\n",
       "      <td>100767279</td>\n",
       "      <td>...</td>\n",
       "      <td>22632</td>\n",
       "      <td>24919</td>\n",
       "      <td>453159</td>\n",
       "      <td>534353</td>\n",
       "      <td>611171</td>\n",
       "      <td>717370</td>\n",
       "      <td>7528</td>\n",
       "      <td>100022685</td>\n",
       "      <td>118976930</td>\n",
       "      <td>121434266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZBTB17</th>\n",
       "      <td>100050436</td>\n",
       "      <td>100620358</td>\n",
       "      <td>100348036</td>\n",
       "      <td>100398482</td>\n",
       "      <td>100454123</td>\n",
       "      <td>100482510</td>\n",
       "      <td>100589356</td>\n",
       "      <td>100654604</td>\n",
       "      <td>100731568</td>\n",
       "      <td>100754046</td>\n",
       "      <td>...</td>\n",
       "      <td>22642</td>\n",
       "      <td>313666</td>\n",
       "      <td>740453</td>\n",
       "      <td>512915</td>\n",
       "      <td>607083</td>\n",
       "      <td>696327</td>\n",
       "      <td>7709</td>\n",
       "      <td>100619919</td>\n",
       "      <td>118986310</td>\n",
       "      <td>121448060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF830</th>\n",
       "      <td>100071676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100344684</td>\n",
       "      <td>100385084</td>\n",
       "      <td>100455596</td>\n",
       "      <td>100483202</td>\n",
       "      <td>100603392</td>\n",
       "      <td>100676058</td>\n",
       "      <td>100715797</td>\n",
       "      <td>100764718</td>\n",
       "      <td>...</td>\n",
       "      <td>66983</td>\n",
       "      <td>497967</td>\n",
       "      <td>454580</td>\n",
       "      <td>539497</td>\n",
       "      <td>480608</td>\n",
       "      <td>715130</td>\n",
       "      <td>91603</td>\n",
       "      <td>107651029</td>\n",
       "      <td>119001290</td>\n",
       "      <td>121456534</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>681 rows × 201 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Equus_caballus Sus_scrofa Oryctolagus_cuniculus Callithrix_jacchus  \\\n",
       "AAAS          100063811  100154333             100356621          100405154   \n",
       "ABL1          100069715  100524544             100345909          100398414   \n",
       "ABRAXAS1      100061483  100524450             100350582          100401953   \n",
       "ACADVL        100061583  100520636             100346654          100399197   \n",
       "ACD           100066184  100623181             100346875          100405339   \n",
       "...                 ...        ...                   ...                ...   \n",
       "YIF1A         100052354  100521064             127489740          100394171   \n",
       "YWHAE         100059962  100620600                   NaN          100400280   \n",
       "YY1           100630389  102162035             103346032          100411814   \n",
       "ZBTB17        100050436  100620358             100348036          100398482   \n",
       "ZNF830        100071676        NaN             100344684          100385084   \n",
       "\n",
       "         Pongo_abelii Ailuropoda_melanoleuca Nomascus_leucogenys  \\\n",
       "AAAS        100444879              100468100           100586295   \n",
       "ABL1        100453216              100466468           100601002   \n",
       "ABRAXAS1    100449014              100463877           100597649   \n",
       "ACADVL      100454381              100472785           100585921   \n",
       "ACD         100447506              100474125           100586139   \n",
       "...               ...                    ...                 ...   \n",
       "YIF1A       100443779              100482492           100597699   \n",
       "YWHAE       100455220              100484851           100600918   \n",
       "YY1         100447742              100474192           100583435   \n",
       "ZBTB17      100454123              100482510           100589356   \n",
       "ZNF830      100455596              100483202           100603392   \n",
       "\n",
       "         Loxodonta_africana Cavia_porcellus Cricetulus_griseus  ...  \\\n",
       "AAAS              100662257       100727054          100767976  ...   \n",
       "ABL1              100677152       100720976          100768098  ...   \n",
       "ABRAXAS1          100675954       100721329          100766855  ...   \n",
       "ACADVL            100657276       100713909          100767527  ...   \n",
       "ACD               100661748       100730199          100763518  ...   \n",
       "...                     ...             ...                ...  ...   \n",
       "YIF1A             100656055       100727936          100763610  ...   \n",
       "YWHAE             100658987       100714396                NaN  ...   \n",
       "YY1               100673864       100730873          100767279  ...   \n",
       "ZBTB17            100654604       100731568          100754046  ...   \n",
       "ZNF830            100676058       100715797          100764718  ...   \n",
       "\n",
       "         Mus_musculus Rattus_norvegicus Pan_troglodytes Bos_taurus  \\\n",
       "AAAS           223921            300259          467003     506561   \n",
       "ABL1            11350            311860       107976801     540876   \n",
       "ABRAXAS1        70681            289468          461205     504796   \n",
       "ACADVL          11370             25363          455237     282130   \n",
       "ACD            497652            307798          454170     510353   \n",
       "...               ...               ...             ...        ...   \n",
       "YIF1A           68090            171441          746258     506336   \n",
       "YWHAE           22627             29753          454409     282125   \n",
       "YY1             22632             24919          453159     534353   \n",
       "ZBTB17          22642            313666          740453     512915   \n",
       "ZNF830          66983            497967          454580     539497   \n",
       "\n",
       "         Canis_lupus_familiaris Macaca_mulatta Homo_sapiens  \\\n",
       "AAAS                     607867         719095         8086   \n",
       "ABL1                     491292         722449           25   \n",
       "ABRAXAS1                 478459         713263        84142   \n",
       "ACADVL                   489463         714326           37   \n",
       "ACD                      487931         699230        65057   \n",
       "...                         ...            ...          ...   \n",
       "YIF1A                    476014         714486        10897   \n",
       "YWHAE                    480645         720156         7531   \n",
       "YY1                      611171         717370         7528   \n",
       "ZBTB17                   607083         696327         7709   \n",
       "ZNF830                   480608         715130        91603   \n",
       "\n",
       "         Monodelphis_domestica Sturnira_hondurensis Microtus_oregoni  \n",
       "AAAS                       NaN                  NaN              NaN  \n",
       "ABL1                 100619549            118980343        121462621  \n",
       "ABRAXAS1             100017569            118979544        121440920  \n",
       "ACADVL               100016619            118974685        121456541  \n",
       "ACD                  103105155            118983853        121447226  \n",
       "...                        ...                  ...              ...  \n",
       "YIF1A                100029722            118991769        121454377  \n",
       "YWHAE                100012276            119003777        121456736  \n",
       "YY1                  100022685            118976930        121434266  \n",
       "ZBTB17               100619919            118986310        121448060  \n",
       "ZNF830               107651029            119001290        121456534  \n",
       "\n",
       "[681 rows x 201 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes_organisms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 3 Animal Traits & AnAge databases; calculating longevity quotient\n",
    "\n",
    "**Description of script:** Edits data from Animal Traits dataset and saves name of an organism and its body mass to table 'df_animaltraits'. Merges AnAge data and Animal Traits data (table 'df_anage_animaltraits'); and calculate longevity quotient. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Script: Load Animal Traits dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_animaltraits = pd.read_csv(r'{0}\\AnimalTraits_dataset\\AnimalTraits.csv'.format(main_path), sep=';')\n",
    "\n",
    "animaltraits_bodymass = {}\n",
    "\n",
    "for index, rows in df_animaltraits.iterrows():\n",
    "        name = str(rows.species)\n",
    "        if 'e_' in str(rows.body_mass): \n",
    "            main_num = str(rows.body_mass).split('e_0')\n",
    "            body_mass = float(main_num[0]) / (10 ** float(main_num[1])) * 1000 \n",
    "            df_animaltraits.at[index,'body_mass'] = body_mass\n",
    "            \n",
    "        else: \n",
    "            body_mass = round(float(rows.body_mass) * 1000, 2)\n",
    "            df_animaltraits.at[index,'body_mass'] = body_mass\n",
    "        animaltraits_bodymass.update({name: body_mass})\n",
    "        \n",
    "df_animaltraits = df_animaltraits.iloc[:, [5, 12]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Results: List of organisms and their body mass (in g) from Animal Traits dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>body_mass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hyla_versicolor</td>\n",
       "      <td>13.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pseudacris_triseriata</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acris_crepitans</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Platymantis_vitiensis</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rana_sylvatica</td>\n",
       "      <td>0.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3575</th>\n",
       "      <td>Sphaerodactylus_cinereus</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3576</th>\n",
       "      <td>Sphaerodactylus_notatus</td>\n",
       "      <td>0.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>Anolis_sagrei</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3578</th>\n",
       "      <td>Coleonyx_switaki</td>\n",
       "      <td>9.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3579</th>\n",
       "      <td>Coleonyx_variegatus</td>\n",
       "      <td>3.62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3580 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       species body_mass\n",
       "0              Hyla_versicolor     13.15\n",
       "1        Pseudacris_triseriata       0.1\n",
       "2              Acris_crepitans       0.3\n",
       "3        Platymantis_vitiensis       0.3\n",
       "4               Rana_sylvatica      0.37\n",
       "...                        ...       ...\n",
       "3575  Sphaerodactylus_cinereus      0.54\n",
       "3576   Sphaerodactylus_notatus      0.33\n",
       "3577             Anolis_sagrei       0.4\n",
       "3578          Coleonyx_switaki      9.46\n",
       "3579       Coleonyx_variegatus      3.62\n",
       "\n",
       "[3580 rows x 2 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_animaltraits"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Fill in missing body masses in AnAge with data from Animal Traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Sort(sub_li):\n",
    "    return(sorted(sub_li, key = lambda x: x[1]))   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_anage = pd.read_csv(r'{0}\\AnAge_dataset\\anage_data.csv'.format(main_path), sep=';')\n",
    "organisms_downloaded = df_genes_organisms.columns.values.tolist()\n",
    "  \n",
    "for index, rows in df_anage.iterrows():\n",
    "        name = str(rows.Genus) + '_' + str(rows.Species)\n",
    "        if math.isnan(rows.Body_mass):\n",
    "            if name in animaltraits_bodymass.keys():\n",
    "                df_anage.at[index,'Body_mass'] = animaltraits_bodymass[name]\n",
    "                \n",
    "df_anage_animaltraits = df_anage.iloc[:, [6,7,20,28]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Results: AnAge and Animal Traits merged; table containing max. longevity and body mass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Maximum_longevity</th>\n",
       "      <th>Body_mass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Daphnia</td>\n",
       "      <td>pulicaria</td>\n",
       "      <td>0.19</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Drosophila</td>\n",
       "      <td>melanogaster</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.00095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Apis</td>\n",
       "      <td>mellifera</td>\n",
       "      <td>8.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cardiocondyla</td>\n",
       "      <td>obscurior</td>\n",
       "      <td>0.50</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Lasius</td>\n",
       "      <td>niger</td>\n",
       "      <td>28.00</td>\n",
       "      <td>0.00058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4214</th>\n",
       "      <td>Scolymastra</td>\n",
       "      <td>joubini</td>\n",
       "      <td>15000.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4215</th>\n",
       "      <td>Pinus</td>\n",
       "      <td>longaeva</td>\n",
       "      <td>5062.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4216</th>\n",
       "      <td>Saccharomyces</td>\n",
       "      <td>cerevisiae</td>\n",
       "      <td>0.04</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4217</th>\n",
       "      <td>Schizosaccharomyces</td>\n",
       "      <td>pombe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4218</th>\n",
       "      <td>Podospora</td>\n",
       "      <td>anserina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4219 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Genus       Species  Maximum_longevity  Body_mass\n",
       "0                 Daphnia     pulicaria               0.19        NaN\n",
       "1              Drosophila  melanogaster               0.30    0.00095\n",
       "2                    Apis     mellifera               8.00        NaN\n",
       "3           Cardiocondyla     obscurior               0.50        NaN\n",
       "4                  Lasius         niger              28.00    0.00058\n",
       "...                   ...           ...                ...        ...\n",
       "4214          Scolymastra       joubini           15000.00        NaN\n",
       "4215                Pinus      longaeva            5062.00        NaN\n",
       "4216        Saccharomyces    cerevisiae               0.04        NaN\n",
       "4217  Schizosaccharomyces         pombe                NaN        NaN\n",
       "4218            Podospora      anserina                NaN        NaN\n",
       "\n",
       "[4219 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_anage_animaltraits"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script: Calculate longevity quotient and choose low and top decile (aka short and long-living species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "organisms_in_organisms_lq_list =[]\n",
    "organisms_lq_list_all = []\n",
    "organisms_lq_list_final = []\n",
    "\n",
    "# Go through the AnAge + Animal Traits table and calculate longevity quotient for each organism\n",
    "\n",
    "for index, rows in df_anage_animaltraits.iterrows():\n",
    "        name = str(rows.Genus) + '_' + str(rows.Species)\n",
    "        if not math.isnan(rows.Body_mass) and not math.isnan(rows.Maximum_longevity):\n",
    "           \n",
    "            # maximum longevity\n",
    "            ml = float(rows.Maximum_longevity)\n",
    "            \n",
    "            # body mass\n",
    "            bm = float(rows.Body_mass)\n",
    "\n",
    "            # longevity quotient\n",
    "            lq = round(ml/(6.32*(bm**0.139)), 2)\n",
    "            \n",
    "            \n",
    "            \n",
    "            # All organisms from AnAge and Animal Traits datasets that data on body mass and lifespan exist\n",
    "            \n",
    "            organisms_lq_list_all.append({'organism':name, 'maximum_lifespan':ml, 'body_mass':bm, 'longevity_quotient':lq})\n",
    "            \n",
    "            \n",
    "            \n",
    "            # Overlap between list of organisms, that we have data for, and list of organisms with known bodymasses and lifespans\n",
    "            \n",
    "            if name in organisms_downloaded: \n",
    "                organisms_lq_list_final.append({'organism':name, 'maximum_lifespan':ml, 'body_mass':bm, 'longevity_quotient':lq})\n",
    "        \n",
    "        \n",
    "# FINAL table of all organisms that data for orthologs and longevity quotient are available  \n",
    "            \n",
    "df_organisms_lq = pd.DataFrame.from_dict(organisms_lq_list_final)\n",
    "\n",
    "\n",
    "\n",
    "# Calculate low and top decile of longevity quotients and save them as short and long-living organisms\n",
    "\n",
    "df_organisms_lq['decile_rank'] = pd.qcut(df_organisms_lq['longevity_quotient'], 10, labels = False)\n",
    "\n",
    "df_short_living = df_organisms_lq.loc[df_organisms_lq['decile_rank'].isin([0])]\n",
    "short_living_list = df_short_living['organism'].to_list()\n",
    "\n",
    "df_long_living = df_organisms_lq.loc[df_organisms_lq['decile_rank'].isin([9])]\n",
    "long_living_list = df_long_living['organism'].to_list()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results: List of organisms sorted by decile rank\n",
    "Low decile - short-living organisms\n",
    "\n",
    "Top decile - long-living organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>maximum_lifespan</th>\n",
       "      <th>body_mass</th>\n",
       "      <th>longevity_quotient</th>\n",
       "      <th>decile_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Condylura_cristata</td>\n",
       "      <td>2.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Sorex_araneus</td>\n",
       "      <td>3.2</td>\n",
       "      <td>8.4</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Rattus_rattus</td>\n",
       "      <td>4.2</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Rattus_norvegicus</td>\n",
       "      <td>3.8</td>\n",
       "      <td>206.9</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>4.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Heterocephalus_glaber</td>\n",
       "      <td>31.0</td>\n",
       "      <td>35.3</td>\n",
       "      <td>2.99</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Pan_troglodytes</td>\n",
       "      <td>59.4</td>\n",
       "      <td>43990.0</td>\n",
       "      <td>2.13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>90.0</td>\n",
       "      <td>70000.0</td>\n",
       "      <td>3.02</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Myotis_lucifugus</td>\n",
       "      <td>34.0</td>\n",
       "      <td>5.8</td>\n",
       "      <td>4.21</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Tachyglossus_aculeatus</td>\n",
       "      <td>49.5</td>\n",
       "      <td>2909.0</td>\n",
       "      <td>2.58</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  organism  maximum_lifespan  body_mass  longevity_quotient  \\\n",
       "85      Condylura_cristata               2.5       49.0                0.23   \n",
       "83           Sorex_araneus               3.2        8.4                0.38   \n",
       "77           Rattus_rattus               4.2      117.0                0.34   \n",
       "76       Rattus_norvegicus               3.8      206.9                0.29   \n",
       "75            Mus_musculus               4.0       18.0                0.42   \n",
       "..                     ...               ...        ...                 ...   \n",
       "61   Heterocephalus_glaber              31.0       35.3                2.99   \n",
       "56         Pan_troglodytes              59.4    43990.0                2.13   \n",
       "55            Homo_sapiens              90.0    70000.0                3.02   \n",
       "31        Myotis_lucifugus              34.0        5.8                4.21   \n",
       "44  Tachyglossus_aculeatus              49.5     2909.0                2.58   \n",
       "\n",
       "    decile_rank  \n",
       "85            0  \n",
       "83            0  \n",
       "77            0  \n",
       "76            0  \n",
       "75            0  \n",
       "..          ...  \n",
       "61            9  \n",
       "56            9  \n",
       "55            9  \n",
       "31            9  \n",
       "44            9  \n",
       "\n",
       "[86 rows x 5 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_organisms_lq.sort_values('decile_rank')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results: List of short-living organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>maximum_lifespan</th>\n",
       "      <th>body_mass</th>\n",
       "      <th>longevity_quotient</th>\n",
       "      <th>decile_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Antechinus_flavipes</td>\n",
       "      <td>3.9</td>\n",
       "      <td>46.5</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Monodelphis_domestica</td>\n",
       "      <td>5.1</td>\n",
       "      <td>104.0</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Arvicola_amphibius</td>\n",
       "      <td>2.5</td>\n",
       "      <td>94.7</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Mesocricetus_auratus</td>\n",
       "      <td>3.9</td>\n",
       "      <td>108.2</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Mus_musculus</td>\n",
       "      <td>4.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Rattus_norvegicus</td>\n",
       "      <td>3.8</td>\n",
       "      <td>206.9</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Rattus_rattus</td>\n",
       "      <td>4.2</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Sorex_araneus</td>\n",
       "      <td>3.2</td>\n",
       "      <td>8.4</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Condylura_cristata</td>\n",
       "      <td>2.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 organism  maximum_lifespan  body_mass  longevity_quotient  \\\n",
       "34    Antechinus_flavipes               3.9       46.5                0.36   \n",
       "36  Monodelphis_domestica               5.1      104.0                0.42   \n",
       "65     Arvicola_amphibius               2.5       94.7                0.21   \n",
       "66   Mesocricetus_auratus               3.9      108.2                0.32   \n",
       "75           Mus_musculus               4.0       18.0                0.42   \n",
       "76      Rattus_norvegicus               3.8      206.9                0.29   \n",
       "77          Rattus_rattus               4.2      117.0                0.34   \n",
       "83          Sorex_araneus               3.2        8.4                0.38   \n",
       "85     Condylura_cristata               2.5       49.0                0.23   \n",
       "\n",
       "    decile_rank  \n",
       "34            0  \n",
       "36            0  \n",
       "65            0  \n",
       "66            0  \n",
       "75            0  \n",
       "76            0  \n",
       "77            0  \n",
       "83            0  \n",
       "85            0  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_short_living"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Results: List of long-living organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organism</th>\n",
       "      <th>maximum_lifespan</th>\n",
       "      <th>body_mass</th>\n",
       "      <th>longevity_quotient</th>\n",
       "      <th>decile_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Desmodus_rotundus</td>\n",
       "      <td>29.2</td>\n",
       "      <td>29.4</td>\n",
       "      <td>2.89</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Pteropus_giganteus</td>\n",
       "      <td>40.0</td>\n",
       "      <td>562.2</td>\n",
       "      <td>2.62</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Eptesicus_fuscus</td>\n",
       "      <td>19.0</td>\n",
       "      <td>13.3</td>\n",
       "      <td>2.10</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Myotis_lucifugus</td>\n",
       "      <td>34.0</td>\n",
       "      <td>5.8</td>\n",
       "      <td>4.21</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Myotis_myotis</td>\n",
       "      <td>37.1</td>\n",
       "      <td>25.0</td>\n",
       "      <td>3.75</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Tachyglossus_aculeatus</td>\n",
       "      <td>49.5</td>\n",
       "      <td>2909.0</td>\n",
       "      <td>2.58</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>90.0</td>\n",
       "      <td>70000.0</td>\n",
       "      <td>3.02</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Pan_troglodytes</td>\n",
       "      <td>59.4</td>\n",
       "      <td>43990.0</td>\n",
       "      <td>2.13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Heterocephalus_glaber</td>\n",
       "      <td>31.0</td>\n",
       "      <td>35.3</td>\n",
       "      <td>2.99</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  organism  maximum_lifespan  body_mass  longevity_quotient  \\\n",
       "24       Desmodus_rotundus              29.2       29.4                2.89   \n",
       "27      Pteropus_giganteus              40.0      562.2                2.62   \n",
       "30        Eptesicus_fuscus              19.0       13.3                2.10   \n",
       "31        Myotis_lucifugus              34.0        5.8                4.21   \n",
       "32           Myotis_myotis              37.1       25.0                3.75   \n",
       "44  Tachyglossus_aculeatus              49.5     2909.0                2.58   \n",
       "55            Homo_sapiens              90.0    70000.0                3.02   \n",
       "56         Pan_troglodytes              59.4    43990.0                2.13   \n",
       "61   Heterocephalus_glaber              31.0       35.3                2.99   \n",
       "\n",
       "    decile_rank  \n",
       "24            9  \n",
       "27            9  \n",
       "30            9  \n",
       "31            9  \n",
       "32            9  \n",
       "44            9  \n",
       "55            9  \n",
       "56            9  \n",
       "61            9  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_long_living"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script: Create dictionaries to choose one seq per geneid for selected long and short-living organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for long-living organisms, all orthologous sequences sorted by geneid and refseq\n",
    "long_orthologs_refseqs_sorted = {} \n",
    "\n",
    "geneids_long = []\n",
    "\n",
    "for organism in long_living_list: \n",
    "    geneids_long += df_genes_organisms[organism].to_list()\n",
    "\n",
    "geneids_long = [b for b in geneids_long if isinstance(b, str)] # remove all nan\n",
    "\n",
    "for geneid in geneids_long:\n",
    "     long_orthologs_refseqs_sorted.update({geneid:all_orthologs_refseqs_sorted[geneid]})\n",
    "    \n",
    "\n",
    "# for short-living organisms, all orthologous sequences sorted by geneid and refseq\n",
    "short_orthologs_refseqs_sorted = {}\n",
    "\n",
    "geneids_short = []\n",
    "\n",
    "for organism in short_living_list:\n",
    "    geneids_short += df_genes_organisms[organism].to_list()\n",
    "\n",
    "geneids_short = [b for b in geneids_short if isinstance(b, str)] # remove all nan\n",
    "\n",
    "for geneid in geneids_short:\n",
    "    short_orthologs_refseqs_sorted.update({geneid:all_orthologs_refseqs_sorted[geneid]})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 4 SHORT-living; Select reference sequences \n",
    "**Description of script:** It chooses which protein sequence will be used as a reference sequence.\n",
    "\n",
    "Sequence is chosen as a reference if there are keywords 'RefSeq Select' in GenPept file of the protein - script asks NCBI API for GenPept file and check it. Else the longest sequence from all protein sequences of that gene is chosen.\n",
    "\n",
    "seq_selection: \n",
    "  1. 'only_one_seq'; there was only one reference sequence in NCBI dtb\n",
    "  2. 'RefSeq_Select'; RefSeq Select in keywords of GenPept file\n",
    "  3. 'longest_seq'; the longest protein sequence\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Variables and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "short_selected_orthologs_list = [] \n",
    "# = list made of the 'orthologs_list' but containing only one seq per gene id\n",
    "# + information if it was chosen with RefSeq Select keyword or it is the longest sequence\n",
    "# [{human_gene_name: X, gene id: X, gene name: X, organism name; X, accession number: X, protein sequence: X, seq_selection: X}, {...}]\n",
    "\n",
    "\n",
    "# testing dictionary \n",
    "# short_orthologs_refseqs_sorted = {'gene_id1': {'NP_001104745': {'protein_sequence':'X', 'human_gene_name': 'X', 'gene_name': 'X', 'organism': 'X'}, 'XP_023505037.1': {'protein_sequence':'X', 'human_gene_name': 'X', 'gene_name': 'X', 'organism': 'X'}}, 'gene_id2': {'XP_023505041.1' :{'protein_sequence':'X', 'human_gene_name': 'X', 'gene_name': 'X', 'organism': 'X'}}}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunks(data, SIZE=10000):\n",
    "    it = iter(data)\n",
    "    for i in range(0, len(data), SIZE):\n",
    "        yield {k:data[k] for k in islice(it, SIZE)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Parse dictionary short_orthologs_refseqs_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#all refseqs from dictionary short_orthologs_refseqs_sorted\n",
    "\n",
    "all_refseqs_list = [] \n",
    "gene_ids_list = []\n",
    "\n",
    "for key in list(short_orthologs_refseqs_sorted.keys()):\n",
    "    gene_ids_list.append(key)\n",
    "    for jey in list(short_orthologs_refseqs_sorted[key].keys()): \n",
    "        all_refseqs_list.append(jey)\n",
    "        \n",
    "print('Count of all gene ids:', len(gene_ids_list))\n",
    "print('Count of all refseqs:', len(all_refseqs_list), '\\n')\n",
    "\n",
    "\n",
    "\n",
    "# split the dictionary by 50 \n",
    "splitted_short_orthologs_refseqs_sorted = []\n",
    "\n",
    "\n",
    "for item in chunks({i:short_orthologs_refseqs_sorted[i] for i in short_orthologs_refseqs_sorted}, 10):\n",
    "    splitted_short_orthologs_refseqs_sorted.append(item)\n",
    "    \n",
    "    \n",
    "count_list = []\n",
    "count = 0\n",
    "splitted_refseqs_list = []\n",
    "for item in splitted_short_orthologs_refseqs_sorted:\n",
    "    for key in list(item.keys()):\n",
    "        \n",
    "        for jey in list(item[key].keys()):\n",
    "            count+=1\n",
    "            splitted_refseqs_list.append(jey)\n",
    "    count_list.append(count)\n",
    "    count = 0\n",
    "\n",
    "print('Count in splitted:', len(splitted_refseqs_list))\n",
    "\n",
    "\n",
    "print('Number of sub-parts of short_orthologs_refseqs_sorted to be examined:', len(splitted_short_orthologs_refseqs_sorted))\n",
    "# print('Counts of refseqs in subparts:', count_list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Selection - one protein seq per gene id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for chunk in splitted_short_orthologs_refseqs_sorted:\n",
    "    \n",
    "    \n",
    "    refseqs_string = '' \n",
    "    refseqs_list = []\n",
    "    \n",
    "    \n",
    "    for key in list(chunk.keys()):\n",
    "        for jey in list(chunk[key].keys()): \n",
    "            refseqs_string = refseqs_string + '&id=' + jey\n",
    "            refseqs_list.append(jey)\n",
    "                    \n",
    "    \n",
    "    urls = requests.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein{}&api_key=aabfba1a0deb347c0ffdd37c6e46908c6a09&rettype=gp\".format(refseqs_string), auth=('user', 'pass'), timeout=(3.05, 27))\n",
    "    genpept_files = urls.text.split('//')\n",
    "    if '\\n\\n' in genpept_files: genpept_files.remove('\\n\\n')\n",
    "\n",
    "    # print('Number of downloaded files is the same as number of RefSeq numbers:', len(genpept_files) == len(refseqs_list))\n",
    "    # print(len(genpept_files), len(refseqs_list))\n",
    "    \n",
    "    count = -1\n",
    "    for geneid, refseqs_of_geneid in chunk.items(): \n",
    "        refseq_select = None\n",
    "        the_longest_seq = None\n",
    "        seq_len = 0\n",
    "\n",
    "        refseqs = list(refseqs_of_geneid.keys())\n",
    "      \n",
    "        # if there is more than one RefSeq sequence for gene id\n",
    "        if len(refseqs) > 1 and count < len(genpept_files):\n",
    "            for refseq in refseqs:\n",
    "                count += 1\n",
    "                genpept_file = genpept_files[count]\n",
    "                if refseq not in genpept_file: print('Error: File does not correspond to RefSeq number.')\n",
    "\n",
    "                # if the reference sequence is RefSeq Select, save it\n",
    "                if 'RefSeq Select' in genpept_file:    \n",
    "                    refseq_select = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'RefSeq_Select'}\n",
    "\n",
    "\n",
    "                # save the longest reference sequence \n",
    "                if len(refseqs_of_geneid[refseq]['protein_sequence']) > seq_len: \n",
    "                    seq_len = len(refseqs_of_geneid[refseq]['protein_sequence'])\n",
    "                    the_longest_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'longest_seq'}\n",
    "\n",
    "\n",
    "            if refseq_select: \n",
    "                short_selected_orthologs_list.append(refseq_select)\n",
    "\n",
    "            elif the_longest_seq: \n",
    "                short_selected_orthologs_list.append(the_longest_seq)  \n",
    "            \n",
    "\n",
    "        else: # there is only one RefSeq\n",
    "            count += 1\n",
    "            refseq = list(refseqs_of_geneid.keys())[0]\n",
    "            only_one_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'only_one_seq'}\n",
    "            short_selected_orthologs_list.append(only_one_seq) \n",
    "\n",
    "\n",
    "\n",
    "df_short_selected_orthologs_list = pd.DataFrame.from_dict(short_selected_orthologs_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Results: Short-living organisms; table of selected reference sequences; one per gene id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_short_selected_orthologs_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 5 LONG-living; Select reference sequences \n",
    "**Description of script:** It chooses which protein sequence will be used as a reference sequence.\n",
    "\n",
    "Sequence is chosen as a reference if there are keywords 'RefSeq Select' in GenPept file of the protein - script asks NCBI API for GenPept file and check it. Else the longest sequence from all protein sequences of that gene is chosen.\n",
    "\n",
    "seq_selection: \n",
    "  1. 'only_one_seq'; there was only one reference sequence in NCBI dtb\n",
    "  2. 'RefSeq_Select'; RefSeq Select in keywords of GenPept file\n",
    "  3. 'longest_seq'; the longest protein sequence\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Variables and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "long_selected_orthologs_list = [] \n",
    "# = list made of the 'orthologs_list' but containing only one seq per gene id\n",
    "# + information if it was chosen with RefSeq Select keyword or it is the longest sequence\n",
    "# [{human_gene_name: X, gene id: X, gene name: X, organism name; X, accession number: X, protein sequence: X, seq_selection: X}, {...}]\n",
    "\n",
    "\n",
    "# testing dictionary \n",
    "# all_orthologs_refseqs_sorted = {'gene_id1': {'NP_001104745': {'protein_sequence':'X', 'human_gene_name': 'X', 'gene_name': 'X', 'organism': 'X'}, 'XP_023505037.1': {'protein_sequence':'X', 'human_gene_name': 'X', 'gene_name': 'X', 'organism': 'X'}}, 'gene_id2': {'XP_023505041.1' :{'protein_sequence':'X', 'human_gene_name': 'X', 'gene_name': 'X', 'organism': 'X'}}}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunks(data, SIZE=10000):\n",
    "    it = iter(data)\n",
    "    for i in range(0, len(data), SIZE):\n",
    "        yield {k:data[k] for k in islice(it, SIZE)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Parse dictionary long_orthologs_refseqs_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#all refseqs from dictionary long_orthologs_refseqs_sorted\n",
    "\n",
    "all_refseqs_list = [] \n",
    "gene_ids_list = []\n",
    "\n",
    "for key in list(long_orthologs_refseqs_sorted.keys()):\n",
    "    gene_ids_list.append(key)\n",
    "    for jey in list(long_orthologs_refseqs_sorted[key].keys()): \n",
    "        all_refseqs_list.append(jey)\n",
    "        \n",
    "print('Count of all gene ids:', len(gene_ids_list))\n",
    "print('Count of all refseqs:', len(all_refseqs_list), '\\n')\n",
    "\n",
    "\n",
    "\n",
    "# split the dictionary by 50 \n",
    "splitted_long_orthologs_refseqs_sorted = []\n",
    "\n",
    "\n",
    "for item in chunks({i:long_orthologs_refseqs_sorted[i] for i in long_orthologs_refseqs_sorted}, 10):\n",
    "    splitted_long_orthologs_refseqs_sorted.append(item)\n",
    "    \n",
    "    \n",
    "count_list = []\n",
    "count = 0\n",
    "splitted_refseqs_list = []\n",
    "for item in splitted_long_orthologs_refseqs_sorted:\n",
    "    for key in list(item.keys()):\n",
    "        \n",
    "        for jey in list(item[key].keys()):\n",
    "            count+=1\n",
    "            splitted_refseqs_list.append(jey)\n",
    "    count_list.append(count)\n",
    "    count = 0\n",
    "\n",
    "print('Count in splitted:', len(splitted_refseqs_list))\n",
    "\n",
    "\n",
    "print('Number of sub-parts of long_orthologs_refseqs_sorted to be examined:', len(splitted_long_orthologs_refseqs_sorted))\n",
    "# print('Counts of refseqs in subparts:', count_list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: Selection - one protein seq per gene id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for chunk in splitted_long_orthologs_refseqs_sorted:\n",
    "    \n",
    "    refseqs_string = '' \n",
    "    refseqs_list = []\n",
    "    \n",
    "    \n",
    "    for key in list(chunk.keys()):\n",
    "        for jey in list(chunk[key].keys()): \n",
    "            refseqs_string = refseqs_string + '&id=' + jey\n",
    "            refseqs_list.append(jey)\n",
    "                    \n",
    "    \n",
    "    urls = requests.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein{}&api_key=aabfba1a0deb347c0ffdd37c6e46908c6a09&rettype=gp\".format(refseqs_string), auth=('user', 'pass'), timeout=(10, 27))\n",
    "    genpept_files = urls.text.split('//')\n",
    "    if '\\n\\n' in genpept_files: genpept_files.remove('\\n\\n')\n",
    "    print(genpept_files)\n",
    "\n",
    "    # print('Number of downloaded files is the same as number of RefSeq numbers:', len(genpept_files) == len(refseqs_list))\n",
    "    # print(len(genpept_files), len(refseqs_list))\n",
    "\n",
    "    \n",
    "    count = -1\n",
    "    for geneid, refseqs_of_geneid in chunk.items(): \n",
    "        refseq_select = None\n",
    "        the_longest_seq = None\n",
    "        seq_len = 0\n",
    "\n",
    "        refseqs = list(refseqs_of_geneid.keys())\n",
    "      \n",
    "        # if there is more than one RefSeq sequence for gene id\n",
    "        if len(refseqs) > 1 and count < len(genpept_files):\n",
    "            for refseq in refseqs:\n",
    "                count += 1\n",
    "                print(refseq)\n",
    "                print(count)\n",
    "                genpept_file = genpept_files[count]\n",
    "                print(genpept_file)\n",
    "                if refseq not in genpept_file: print('Error: File does not correspond to RefSeq number.')\n",
    "\n",
    "                # if the reference sequence is RefSeq Select, save it\n",
    "                if 'RefSeq Select' in genpept_file:    \n",
    "                    refseq_select = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'RefSeq_Select'}\n",
    "\n",
    "\n",
    "                # save the longest reference sequence \n",
    "                if len(refseqs_of_geneid[refseq]['protein_sequence']) > seq_len: \n",
    "                    seq_len = len(refseqs_of_geneid[refseq]['protein_sequence'])\n",
    "                    the_longest_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'longest_seq'}\n",
    "\n",
    "\n",
    "            if refseq_select: \n",
    "                long_selected_orthologs_list.append(refseq_select)\n",
    "\n",
    "            elif the_longest_seq: \n",
    "                long_selected_orthologs_list.append(the_longest_seq)  \n",
    "            \n",
    "\n",
    "        else: # there is only one RefSeq\n",
    "            count += 1\n",
    "            refseq = list(refseqs_of_geneid.keys())[0]\n",
    "            only_one_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'only_one_seq'}\n",
    "            long_selected_orthologs_list.append(only_one_seq) \n",
    "\n",
    "\n",
    "\n",
    "df_long_selected_orthologs_list = pd.DataFrame.from_dict(long_selected_orthologs_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Results: Long-living organisms; table of selected reference sequences; one per gene id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_long_selected_orthologs_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 6 Create files for CAASTools Discovery analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script: Create config file with organisms and 0 (short-living) / 1 (long-living)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = open(r'{0}\\CAASTools_files\\config.txt'.format(main_path),\"w\")\n",
    "\n",
    "for organism in short_living_list: \n",
    "    config_file.write(organism.lower() + '\\t' + '0' + '\\n')\n",
    "for organism in long_living_list: \n",
    "    config_file.write(organism.lower() + '\\t' + '1' + '\\n')\n",
    "\n",
    "config_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Script: Create fasta file for clustal omega multiple sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_seq = long_selected_orthologs_list + short_selected_orthologs_list\n",
    "genes = df_genes_organisms.index.values.tolist()\n",
    "\n",
    "for gene in genes:\n",
    "    gene_file = open(r'{0}\\CAASTools_files\\fasta_files\\{1}.fasta'.format(main_path, gene),\"w\")\n",
    "\n",
    "    for item in all_seq:\n",
    "        if gene in item.values(): \n",
    "            gene_file.write('>' + str(item['organism']).replace(' ', '_').lower() + '\\n')\n",
    "            gene_file.write(str(item['protein_sequence']) + '\\n')\n",
    "\n",
    "    gene_file.close"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Script: MSAs by Clustal omega\n",
    "!!! It is needed to download the Clustal Omega program and save the folder \"clustal-omega-1.2.2-win64\" with the program \"clustalo.exe\" to main folder "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in genes:\n",
    "    os.system('{0}\\\\clustal-omega-1.2.2-win64\\\\clustalo.exe -i {0}\\\\CAASTools_files\\\\fasta_files\\\\{1}.fasta --out {0}\\\\CAASTools_files\\\\MSA_results\\\\{1}.msa --outfmt clu --wrap=10000'.format(win_path, gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 7 Create csv files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### stress_molecules.csv\n",
    "table of stress molecules (downloaded from Reactome - with Reactome IDs requested, and checked with HGNC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stress_molecules.to_csv(r'{0}\\stress_molecules\\stress_molecules.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### multiple_paths_check.csv\n",
    "table of gene symbols and Reactome IDs; i.e. some genes may be involved in multiple paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiple_paths_df.to_csv(r'{0}\\results\\gene_in_multiple_paths_check.csv'.format(main_path).format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### orthologs.csv \n",
    "table of orthologs and their atributes (i.e. FASTA file parsed into table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_orthologs_list.to_csv(r'{0}\\NCBI_Datasets_download\\orthologs.csv'.format(main_path), sep=',', encoding='utf-8')      "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### orthologs_count.csv\n",
    "table of gene ids and their protein sequence counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ortholog_count_list.to_csv(r'{0}\\results\\orthologs_count.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### genes_organisms_check.csv\n",
    "table of genes-organisms check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_organisms.to_csv(r'{0}\\results\\genes_organisms_check.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### anage_animaltraits_data_merge.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_anage_animaltraits.to_csv(r'{0}\\results\\anage_animaltraits_data_merge.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### organisms_lq.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_organisms_lq.to_csv(r'{0}\\results\\organisms_lq.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### short_living_organisms.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_short_living.to_csv(r'{0}\\results\\short_living_organisms.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### long_living_organisms.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_long_living.to_csv(r'{0}\\results\\long_living_organisms.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### long_selected_orthologs.csv\n",
    "table of selected sequences one reference sequence per gene id (i.e. orthologous gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_long_selected_orthologs_list.to_csv(r'{0}\\results\\long_selected_orthologs.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### short_selected_orthologs.csv\n",
    "table of selected sequences one reference sequence per gene id (i.e. orthologous gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_short_selected_orthologs_list.to_csv(r'{0}\\results\\short_selected_orthologs.csv'.format(main_path), sep=',', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8 Additional: Divide files from NCBI Datasets into subparts and select ref seq if NCBI connection error"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Divide NCBI downloaded files into folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = os.listdir(r'{0}\\NCBI_Datasets_download'.format(main_path))\n",
    "if '.ipynb_checkpoints' in folder: folder.remove('.ipynb_checkpoints')\n",
    "name = 1\n",
    "counter = 0\n",
    "\n",
    "\n",
    "os.system('mkdir {0}\\\\NCBI_Datasets_download\\\\{1}'.format(win_path, str(name)))\n",
    "          \n",
    "for gene in folder:\n",
    "    if counter < 20:\n",
    "        os.system('move {0}\\\\NCBI_Datasets_download\\\\{1} {0}\\\\NCBI_Datasets_download\\\\{2}'.format(win_path, gene, str(name)))\n",
    "        counter += 1\n",
    "        \n",
    "    else:\n",
    "        name += 1\n",
    "        os.system('mkdir {0}\\\\NCBI_Datasets_download\\\\{1}'.format(win_path, str(name)))\n",
    "        counter = 0\n",
    "        os.system('move {0}\\\\NCBI_Datasets_download\\\\{1} {0}\\\\NCBI_Datasets_download\\\\{2}'.format(win_path, gene, str(name)))\n",
    "        counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select reference sequence of every gene for short and long living organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "short_orthologs_refseqs_sorted = {}\n",
    "long_orthologs_refseqs_sorted = {}\n",
    "short_selected_orthologs_list = [] \n",
    "long_selected_orthologs_list = []\n",
    "\n",
    "def most_frequent_count(List):\n",
    "        occurence_count = Counter(List)\n",
    "        return occurence_count.most_common(1)[0][1]\n",
    "\n",
    "def most_frequent_name(List):\n",
    "        occurence_count = Counter(List)\n",
    "        return occurence_count.most_common(1)[0][0]\n",
    "\n",
    "def chunks(data, SIZE=10000):\n",
    "        it = iter(data)\n",
    "        for i in range(0, len(data), SIZE):\n",
    "            yield {k:data[k] for k in islice(it, SIZE)}\n",
    "            \n",
    "df_genes_organisms = pd.read_csv(r'{0}\\analyses\\genes_organisms_check.csv'.format(main_path), sep=',')\n",
    "df_short_living = pd.read_csv(r'{0}\\analyses\\short_living_organisms.csv'.format(main_path), sep=',')\n",
    "short_living_list = df_short_living['organism'].to_list()\n",
    "df_long_living = pd.read_csv(r'{0}analyses\\long_living_organisms.csv'.format(main_path), sep=',')\n",
    "long_living_list = df_long_living['organism'].to_list()\n",
    "\n",
    "\n",
    "folders = os.listdir(r'{0}\\NCBI_Datasets_download\\'.format(main_path))\n",
    "if '.ipynb_checkpoints'in folders: folders.remove('.ipynb_checkpoints')\n",
    "\n",
    "\n",
    "for folder in folders:\n",
    "    print(folder)\n",
    "\n",
    "    # FOR SELECTION OF ONE SEQUENCE PER GENE ID\n",
    "    # {gene_id1: \n",
    "    #     {RefSeq1: \n",
    "    #          {protein_sequence: X, human_gene_name: X, gene_name: X, organism: X}, \n",
    "    #      RefSeq2: {...}\n",
    "    #     }, \n",
    "    #  gene_id2: \n",
    "    #      {RefSeq3: {}\n",
    "    #     }, \n",
    "    #  ...}\n",
    "\n",
    "    \n",
    "\n",
    "    #### Script: Parse FASTA files and save the information into lists and dictionaries\n",
    "    # parse all FASTA files downloaded \n",
    "    # every file is named as the gene which orthologs seqs it contains of\n",
    "    \n",
    "    files = os.listdir(r'{0}\\NCBI_Datasets_download\\{1}'.format(main_path, folder))\n",
    "    if '.ipynb_checkpoints' in files:\n",
    "        files.remove('.ipynb_checkpoints')\n",
    "\n",
    "\n",
    "    print(files)\n",
    "    \n",
    "    all_orthologs_refseqs_sorted = {}\n",
    "\n",
    "    for file in files:\n",
    "        already_been_there = []\n",
    "\n",
    "\n",
    "        gene_name = file.replace('.faa', '')\n",
    "\n",
    "    \n",
    "        with open(r'{0}\\NCBI_Datasets_download\\{1}\\{2}'.format(main_path, folder, file)) as fasta_file:\n",
    "            parser = fastaparser.Reader(fasta_file)\n",
    "\n",
    "            for seq in parser: # for every record (ortholog) of FASTA file        \n",
    "\n",
    "                # split the FASTA record \n",
    "                working_list = seq.description.split(\" [\")     \n",
    "                gene_id = (re.findall('\\\\d+', working_list[2])[0])\n",
    "                # gene_ids.append(gene_id)\n",
    "\n",
    "                organism_name = working_list[1][9:].replace(\"]\", \"\")\n",
    "                organism_name = organism_name.replace(\" \", \"_\")\n",
    "                \n",
    "\n",
    "\n",
    "            \n",
    "                if gene_id not in already_been_there: \n",
    "                    first_sequence = seq.sequence_as_string()\n",
    "                    all_orthologs_refseqs_sorted[gene_id] = {}\n",
    "                    all_orthologs_refseqs_sorted[gene_id].update({seq.id: {'protein_sequence': seq.sequence_as_string(), 'human_gene_name': gene_name, 'gene_name': working_list[0], 'organism': organism_name}})\n",
    "\n",
    "                else:\n",
    "                    another_sequence = seq.sequence_as_string()\n",
    "                    all_orthologs_refseqs_sorted[gene_id].update({seq.id: {'protein_sequence': seq.sequence_as_string(), 'human_gene_name': gene_name, 'gene_name': working_list[0], 'organism': organism_name}})\n",
    "\n",
    "                already_been_there.append(gene_id)\n",
    "\n",
    "    \n",
    "    # SHORT; \n",
    "    # 1. all orthologous sequences sorted by geneid and refseq\n",
    "    short_orthologs_refseqs_sorted = {}\n",
    "\n",
    "    geneids_short = []\n",
    "\n",
    "    for organism in short_living_list:\n",
    "        geneids_short += df_genes_organisms[organism].to_list()\n",
    "\n",
    "    for i in range(len(geneids_short)): \n",
    "        if not math.isnan(geneids_short[i]):\n",
    "            geneids_short[i] = str(math.ceil(geneids_short[i]))\n",
    "        \n",
    "    geneids_short = [b for b in geneids_short if isinstance(b, str)] # remove all nan\n",
    "\n",
    "    for geneid in geneids_short:\n",
    "        if geneid in all_orthologs_refseqs_sorted.keys():\n",
    "            short_orthologs_refseqs_sorted.update({geneid:all_orthologs_refseqs_sorted[geneid]})\n",
    "    \n",
    "    \n",
    "    ### 2. Parse dictionary short_orthologs_refseqs_sorted\n",
    "    #all refseqs from dictionary short_orthologs_refseqs_sorted\n",
    "\n",
    "    all_refseqs_list = [] \n",
    "    gene_ids_list = []\n",
    "\n",
    "    for key in list(short_orthologs_refseqs_sorted.keys()):\n",
    "        gene_ids_list.append(key)\n",
    "        for jey in list(short_orthologs_refseqs_sorted[key].keys()): \n",
    "            all_refseqs_list.append(jey)\n",
    "\n",
    "    print('Count of all gene ids:', len(gene_ids_list))\n",
    "    print('Count of all refseqs:', len(all_refseqs_list), '\\n')\n",
    "\n",
    "\n",
    "\n",
    "    # 3. split the dictionary by 50 \n",
    "    splitted_short_orthologs_refseqs_sorted = []\n",
    "\n",
    "\n",
    "    for item in chunks({i:short_orthologs_refseqs_sorted[i] for i in short_orthologs_refseqs_sorted}, 10):\n",
    "        splitted_short_orthologs_refseqs_sorted.append(item)\n",
    "\n",
    "\n",
    "    count_list = []\n",
    "    count = 0\n",
    "    splitted_refseqs_list = []\n",
    "    for item in splitted_short_orthologs_refseqs_sorted:\n",
    "        for key in list(item.keys()):\n",
    "\n",
    "            for jey in list(item[key].keys()):\n",
    "                count+=1\n",
    "                splitted_refseqs_list.append(jey)\n",
    "        count_list.append(count)\n",
    "        count = 0\n",
    "\n",
    "    print('Count in splitted:', len(splitted_refseqs_list))\n",
    "\n",
    "\n",
    "    print('Number of sub-parts of short_orthologs_refseqs_sorted to be examined:', len(splitted_short_orthologs_refseqs_sorted))\n",
    "    # print('Counts of refseqs in subparts:', count_list)\n",
    "\n",
    "    ### 4. Selection - one protein seq per gene id\n",
    "\n",
    "    for chunk in splitted_short_orthologs_refseqs_sorted:\n",
    "\n",
    "\n",
    "        refseqs_string = '' \n",
    "        refseqs_list = []\n",
    "\n",
    "\n",
    "        for key in list(chunk.keys()):\n",
    "            for jey in list(chunk[key].keys()): \n",
    "                refseqs_string = refseqs_string + '&id=' + jey\n",
    "                refseqs_list.append(jey)\n",
    "\n",
    "                \n",
    "        session = requests.Session()\n",
    "        retry = Retry(connect=5, backoff_factor=1.5)\n",
    "        adapter = HTTPAdapter(max_retries=retry)\n",
    "        session.mount('http://', adapter)\n",
    "        session.mount('https://', adapter)\n",
    "\n",
    "        genpept_files = session.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein{}&api_key=aabfba1a0deb347c0ffdd37c6e46908c6a09&rettype=gp\".format(refseqs_string))\n",
    "        \n",
    "        \n",
    "        # urls = requests.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein{}&api_key=aabfba1a0deb347c0ffdd37c6e46908c6a09&rettype=gp\".format(refseqs_string), auth=('user', 'pass'), timeout=(20, 27))\n",
    "        genpept_files = genpept_files.text.split('//')\n",
    "        if '\\n\\n' in genpept_files: genpept_files.remove('\\n\\n')\n",
    "\n",
    "        # print('Number of downloaded files is the same as number of RefSeq numbers:', len(genpept_files) == len(refseqs_list))\n",
    "        # print(len(genpept_files), len(refseqs_list))\n",
    "\n",
    "        count = -1\n",
    "        for geneid, refseqs_of_geneid in chunk.items(): \n",
    "            refseq_select = None\n",
    "            the_longest_seq = None\n",
    "            seq_len = 0\n",
    "\n",
    "            refseqs = list(refseqs_of_geneid.keys())\n",
    "\n",
    "            # if there is more than one RefSeq sequence for gene id\n",
    "            if len(refseqs) > 1 and count < len(genpept_files):\n",
    "                for refseq in refseqs:\n",
    "                    count += 1\n",
    "                    genpept_file = genpept_files[count]\n",
    "                    if refseq not in genpept_file: print('Error: File does not correspond to RefSeq number.')\n",
    "\n",
    "                    # if the reference sequence is RefSeq Select, save it\n",
    "                    if 'RefSeq Select' in genpept_file:    \n",
    "                        refseq_select = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'RefSeq_Select'}\n",
    "\n",
    "\n",
    "                    # save the longest reference sequence \n",
    "                    if len(refseqs_of_geneid[refseq]['protein_sequence']) > seq_len: \n",
    "                        seq_len = len(refseqs_of_geneid[refseq]['protein_sequence'])\n",
    "                        the_longest_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'longest_seq'}\n",
    "\n",
    "\n",
    "                if refseq_select: \n",
    "                    short_selected_orthologs_list.append(refseq_select)\n",
    "\n",
    "                elif the_longest_seq: \n",
    "                    short_selected_orthologs_list.append(the_longest_seq)  \n",
    "\n",
    "\n",
    "            else: # there is only one RefSeq\n",
    "                count += 1\n",
    "                refseq = list(refseqs_of_geneid.keys())[0]\n",
    "                only_one_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'only_one_seq'}\n",
    "                short_selected_orthologs_list.append(only_one_seq) \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    ##############################\n",
    "    # LONG\n",
    "    # 1. all orthologous sequences sorted by geneid and refseq\n",
    "    long_orthologs_refseqs_sorted = {} \n",
    "\n",
    "    geneids_long = []\n",
    "\n",
    "    for organism in long_living_list: \n",
    "        geneids_long += df_genes_organisms[organism].to_list()\n",
    "    \n",
    "    for i in range(len(geneids_long)): \n",
    "        if not math.isnan(geneids_long[i]):\n",
    "            geneids_long[i] = str(math.ceil(geneids_long[i]))\n",
    "            \n",
    "    geneids_long = [b for b in geneids_long if isinstance(b, str)] # remove all nan\n",
    "\n",
    "    for geneid in geneids_long:\n",
    "        if geneid in all_orthologs_refseqs_sorted.keys():\n",
    "             long_orthologs_refseqs_sorted.update({geneid:all_orthologs_refseqs_sorted[geneid]})\n",
    "\n",
    "                \n",
    "    ### 2. Parse dictionary long_orthologs_refseqs_sorted\n",
    "    #all refseqs from dictionary long_orthologs_refseqs_sorted\n",
    "\n",
    "    all_refseqs_list = [] \n",
    "    gene_ids_list = []\n",
    "\n",
    "    for key in list(long_orthologs_refseqs_sorted.keys()):\n",
    "        gene_ids_list.append(key)\n",
    "        for jey in list(long_orthologs_refseqs_sorted[key].keys()): \n",
    "            all_refseqs_list.append(jey)\n",
    "\n",
    "    print('Count of all gene ids:', len(gene_ids_list))\n",
    "    print('Count of all refseqs:', len(all_refseqs_list), '\\n')\n",
    "\n",
    "\n",
    "\n",
    "    # 3. split the dictionary by 50 \n",
    "    splitted_long_orthologs_refseqs_sorted = []\n",
    "\n",
    "\n",
    "    for item in chunks({i:long_orthologs_refseqs_sorted[i] for i in long_orthologs_refseqs_sorted}, 10):\n",
    "        splitted_long_orthologs_refseqs_sorted.append(item)\n",
    "\n",
    "\n",
    "    count_list = []\n",
    "    count = 0\n",
    "    splitted_refseqs_list = []\n",
    "    for item in splitted_long_orthologs_refseqs_sorted:\n",
    "        for key in list(item.keys()):\n",
    "\n",
    "            for jey in list(item[key].keys()):\n",
    "                count+=1\n",
    "                splitted_refseqs_list.append(jey)\n",
    "        count_list.append(count)\n",
    "        count = 0\n",
    "\n",
    "    print('Count in splitted:', len(splitted_refseqs_list))\n",
    "\n",
    "\n",
    "    print('Number of sub-parts of long_orthologs_refseqs_sorted to be examined:', len(splitted_long_orthologs_refseqs_sorted))\n",
    "    # print('Counts of refseqs in subparts:', count_list)\n",
    "\n",
    "    ### 4. Selection - one protein seq per gene id\n",
    "    \n",
    "    for chunk in splitted_long_orthologs_refseqs_sorted:\n",
    "\n",
    "        refseqs_string = '' \n",
    "        refseqs_list = []\n",
    "\n",
    "\n",
    "        for key in list(chunk.keys()):\n",
    "            for jey in list(chunk[key].keys()): \n",
    "                refseqs_string = refseqs_string + '&id=' + jey\n",
    "                refseqs_list.append(jey)\n",
    "\n",
    "\n",
    "        session = requests.Session()\n",
    "        retry = Retry(connect=5, backoff_factor=1.5)\n",
    "        adapter = HTTPAdapter(max_retries=retry)\n",
    "        session.mount('http://', adapter)\n",
    "        session.mount('https://', adapter)\n",
    "\n",
    "        genpept_files = session.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein{}&api_key=aabfba1a0deb347c0ffdd37c6e46908c6a09&rettype=gp\".format(refseqs_string))\n",
    "        \n",
    "        \n",
    "        genpept_files = genpept_files.text.split('//')\n",
    "               \n",
    "                \n",
    "        # urls = requests.get(\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein{}&api_key=aabfba1a0deb347c0ffdd37c6e46908c6a09&rettype=gp\".format(refseqs_string), auth=('user', 'pass'), timeout=(20, 27))\n",
    "        # genpept_files = urls.text.split('//')\n",
    "        if '\\n\\n' in genpept_files: genpept_files.remove('\\n\\n')\n",
    "        # print(genpept_files)\n",
    "\n",
    "        # print('Number of downloaded files is the same as number of RefSeq numbers:', len(genpept_files) == len(refseqs_list))\n",
    "        # print(len(genpept_files), len(refseqs_list))\n",
    "\n",
    "\n",
    "        count = -1\n",
    "        for geneid, refseqs_of_geneid in chunk.items(): \n",
    "            refseq_select = None\n",
    "            the_longest_seq = None\n",
    "            seq_len = 0\n",
    "\n",
    "            refseqs = list(refseqs_of_geneid.keys())\n",
    "\n",
    "            # if there is more than one RefSeq sequence for gene id\n",
    "            if len(refseqs) > 1 and count < len(genpept_files):\n",
    "                for refseq in refseqs:\n",
    "                    count += 1\n",
    "                    genpept_file = genpept_files[count]\n",
    "                    if refseq not in genpept_file: print('Error: File does not correspond to RefSeq number.')\n",
    "\n",
    "                    # if the reference sequence is RefSeq Select, save it\n",
    "                    if 'RefSeq Select' in genpept_file:    \n",
    "                        refseq_select = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'RefSeq_Select'}\n",
    "\n",
    "\n",
    "                    # save the longest reference sequence \n",
    "                    if len(refseqs_of_geneid[refseq]['protein_sequence']) > seq_len: \n",
    "                        seq_len = len(refseqs_of_geneid[refseq]['protein_sequence'])\n",
    "                        the_longest_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'longest_seq'}\n",
    "\n",
    "\n",
    "                if refseq_select: \n",
    "                    long_selected_orthologs_list.append(refseq_select)\n",
    "\n",
    "                elif the_longest_seq: \n",
    "                    long_selected_orthologs_list.append(the_longest_seq)  \n",
    "\n",
    "\n",
    "            else: # there is only one RefSeq\n",
    "                count += 1\n",
    "                refseq = list(refseqs_of_geneid.keys())[0]\n",
    "                only_one_seq = {'human_gene_name': refseqs_of_geneid[refseq]['human_gene_name'], 'gene_id': geneid, 'gene_name': refseqs_of_geneid[refseq]['gene_name'], 'organism': refseqs_of_geneid[refseq]['organism'], 'accession_number': refseq, 'protein_sequence': refseqs_of_geneid[refseq]['protein_sequence'], 'seq_selection':'only_one_seq'}\n",
    "                long_selected_orthologs_list.append(only_one_seq) \n",
    "\n",
    "\n",
    "\n",
    "df_long_selected_orthologs_list = pd.DataFrame.from_dict(long_selected_orthologs_list)\n",
    "df_short_selected_orthologs_list = pd.DataFrame.from_dict(short_selected_orthologs_list)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
